--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -90,7 +90,7 @@
 import org.apache.commons.lang3.mutable.MutableBoolean;
 import org.jspecify.annotations.Nullable;
 
-public abstract class Level implements LevelAccessor, AutoCloseable, net.neoforged.neoforge.common.extensions.ILevelExtension {
+public abstract class Level extends net.neoforged.neoforge.attachment.AttachmentHolder implements LevelAccessor, AutoCloseable, net.neoforged.neoforge.common.extensions.ILevelExtension {
     public static final Codec<ResourceKey<Level>> RESOURCE_KEY_CODEC = ResourceKey.codec(Registries.DIMENSION);
     public static final ResourceKey<Level> OVERWORLD = ResourceKey.create(Registries.DIMENSION, Identifier.withDefaultNamespace("overworld"));
     public static final ResourceKey<Level> NETHER = ResourceKey.create(Registries.DIMENSION, Identifier.withDefaultNamespace("the_nether"));
@@ -222,11 +227,36 @@
         } else {
             LevelChunk chunk = this.getChunkAt(pos);
             Block block = blockState.getBlock();
+
+            pos = pos.immutable(); // Forge - prevent mutable BlockPos leaks
+            net.neoforged.neoforge.common.util.BlockSnapshot blockSnapshot = null;
+            if (this.captureBlockSnapshots && !this.isClientSide) {
+                blockSnapshot = net.neoforged.neoforge.common.util.BlockSnapshot.create(this.dimension, this, pos, updateFlags);
+                this.capturedBlockSnapshots.add(blockSnapshot);
+            }
+
             BlockState oldState = chunk.setBlockState(pos, blockState, updateFlags);
             if (oldState == null) {
+                if (blockSnapshot != null) this.capturedBlockSnapshots.remove(blockSnapshot);
                 return false;
             } else {
                 BlockState newState = this.getBlockState(pos);
+
+                if (blockSnapshot == null) { // Don't notify clients or update physics while capturing blockstates
+                    this.markAndNotifyBlock(pos, chunk, oldState, blockState, updateFlags, updateLimit);
+                }
+
+                return true;
+            }
+        }
+    }
+
+    // Split off from original setBlockState(BlockPos, BlockState, int, int) method in order to directly send client and physic updates
+    public void markAndNotifyBlock(BlockPos pos, @Nullable LevelChunk chunk, BlockState oldState, BlockState blockState, int updateFlags, int updateLimit) {
+        Block block = blockState.getBlock();
+        BlockState newState = getBlockState(pos);
+        {
+            {
                 if (newState == blockState) {
                     if (oldState != newState) {
                         this.setBlocksDirty(pos, oldState, newState);
@@ -253,9 +283,8 @@
                     }
 
                     this.updatePOIOnBlockStateChange(pos, oldState, newState);
+                    blockState.onBlockStateChange(this, pos, oldState);
                 }
-
-                return true;
             }
         }
     }
@@ -749,6 +819,7 @@
                 hasEntities.setTrue();
                 return AbortableIterationConsumer.Continuation.ABORT;
             } else {
+                if (false)
                 if (e instanceof EnderDragon enderDragon) {
                     for (EnderDragonPart subEntity : enderDragon.getSubEntities()) {
                         T castSubPart = type.tryCast(subEntity);
@@ -762,6 +833,13 @@
                 return AbortableIterationConsumer.Continuation.CONTINUE;
             }
         });
+        for (net.neoforged.neoforge.entity.PartEntity<?> p : this.dragonParts()) {
+            T t = type.tryCast(p);
+            if (t != null && t.getBoundingBox().intersects(bb) && selector.test(t)) {
+                hasEntities.setTrue();
+                break;
+            }
+        }
         return hasEntities.isTrue();
     }
 
