--- a/net/minecraft/world/entity/monster/piglin/Piglin.java
+++ b/net/minecraft/world/entity/monster/piglin/Piglin.java
@@ -357,7 +357,7 @@
     }
 
     protected void holdInOffHand(ItemStack itemStack) {
-        if (itemStack.is(PiglinAi.BARTERING_ITEM)) {
+        if (itemStack.isPiglinCurrency()) {
             this.setItemSlot(EquipmentSlot.OFFHAND, itemStack);
             this.setGuaranteedDrop(EquipmentSlot.OFFHAND);
         } else {
