--- a/net/minecraft/CrashReport.java
+++ b/net/minecraft/CrashReport.java
@@ -92,18 +87,7 @@
             exception.setStackTrace(this.exception.getStackTrace());
         }
 
-        String var4;
-        try {
-            writer = new StringWriter();
-            printWriter = new PrintWriter(writer);
-            exception.printStackTrace(printWriter);
-            var4 = writer.toString();
-        } finally {
-            IOUtils.closeQuietly((Writer)writer);
-            IOUtils.closeQuietly((Writer)printWriter);
-        }
-
-        return var4;
+        return net.neoforged.neoforge.logging.CrashReportExtender.generateEnhancedStackTrace(exception);
     }
 
     public String getFriendlyReport(ReportType reportType, List<String> extraComments) {
