--- a/net/minecraft/CrashReportCategory.java
+++ b/net/minecraft/CrashReportCategory.java
@@ -140,8 +140,10 @@
         if (full.length <= 0) {
             return 0;
         } else {
-            this.stackTrace = new StackTraceElement[full.length - 3 - nestedOffset];
-            System.arraycopy(full, 3 + nestedOffset, this.stackTrace, 0, this.stackTrace.length);
+            int len = full.length - 3 - nestedOffset;
+            if (len <= 0) len = full.length;
+            this.stackTrace = new StackTraceElement[len];
+            System.arraycopy(full, full.length - len, this.stackTrace, 0, this.stackTrace.length);
             return this.stackTrace.length;
         }
     }
