--- a/net/minecraft/world/inventory/EnchantmentMenu.java
+++ b/net/minecraft/world/inventory/EnchantmentMenu.java
@@ -74,7 +74,7 @@
 
             @Override
             public boolean mayPlace(ItemStack itemStack) {
-                return itemStack.is(Items.LAPIS_LAZULI);
+                return itemStack.is(Items.LAPIS_LAZULI); // Neo: TODO - replace with the tag once we have client tags
             }
 
             @Override
@@ -162,14 +163,11 @@
                     List<EnchantmentInstance> newEnchantment = this.getEnchantmentList(level.registryAccess(), itemStack, buttonId, this.costs[buttonId]);
                     if (!newEnchantment.isEmpty()) {
                         player.onEnchantmentPerformed(itemStack, enchantmentCost);
-                        if (itemStack.is(Items.BOOK)) {
-                            enchantmentItem = itemStack.transmuteCopy(Items.ENCHANTED_BOOK);
-                            this.enchantSlots.setItem(0, enchantmentItem);
-                        }
+                        // Neo: Allow items to transform themselves when enchanted, instead of relying on hardcoded transformations for Items.BOOK
+                        enchantmentItem = itemStack.getItem().applyEnchantments(itemStack, newEnchantment);
+                        this.enchantSlots.setItem(0, enchantmentItem);
+                        net.neoforged.neoforge.common.CommonHooks.onPlayerEnchantItem(player, enchantmentItem, newEnchantment);
 
-                        for (EnchantmentInstance enchantment : newEnchantment) {
-                            enchantmentItem.enchant(enchantment.enchantment(), enchantment.level());
-                        }
 
                         currency.consume(enchantmentCost, player);
                         if (currency.isEmpty()) {
