--- a/net/minecraft/world/item/component/FireworkExplosion.java
+++ b/net/minecraft/world/item/component/FireworkExplosion.java
@@ -141,5 +_,9 @@
         public String getSerializedName() {
             return this.name;
         }
+
+        public static net.neoforged.fml.common.asm.enumextension.ExtensionInfo getExtensionInfo() {
+            return net.neoforged.fml.common.asm.enumextension.ExtensionInfo.nonExtended(FireworkExplosion.Shape.class);
+        }
     }
 }
