--- a/net/minecraft/tags/TagBuilder.java
+++ b/net/minecraft/tags/TagBuilder.java
@@ -7,6 +7,8 @@
 public class TagBuilder implements net.neoforged.neoforge.common.extensions.ITagBuilderExtension {
     private final List<TagEntry> entries = new ArrayList<>();
     private boolean replace = false;
+    /// Neo: Entries to be removed from this tag
+    private final List<TagEntry> removeEntries = new ArrayList<>();
 
     public static TagBuilder create() {
         return new TagBuilder();
