--- a/net/minecraft/client/gui/components/EditBox.java
+++ b/net/minecraft/client/gui/components/EditBox.java
@@ -60,6 +60,7 @@
     private long focusedTime = Util.getMillis();
     private int textX;
     private int textY;
+    private java.util.function.Predicate<String> filter = _ -> true; // Neo: Add back text input filtering
 
     public EditBox(Font font, int width, int height, Component narration) {
         this(font, 0, 0, width, height, narration);
