--- a/net/minecraft/world/level/chunk/storage/SerializableChunkData.java
+++ b/net/minecraft/world/level/chunk/storage/SerializableChunkData.java
@@ -122,7 +151,7 @@
             long[] carvingMask = chunkData.getLongArray("carving_mask").orElse(null);
             Map<Heightmap.Types, long[]> heightmaps = new EnumMap<>(Heightmap.Types.class);
             chunkData.getCompound("Heightmaps").ifPresent(heightmapsTag -> {
-                for (Heightmap.Types type : status.heightmapsAfter()) {
+                for (Heightmap.Types type : status.getChunkSaveHeightmaps()) {
                     heightmapsTag.getLongArray(type.getSerializationKey()).ifPresent(longs -> heightmaps.put(type, longs));
                 }
             });
@@ -185,6 +214,9 @@
                 }
             }
 
+            CompoundTag attachmentData = chunkData.getCompound(net.neoforged.neoforge.attachment.AttachmentHolder.ATTACHMENTS_NBT_KEY).orElse(null);
+            ListTag auxLightData = chunkData.getList(net.neoforged.neoforge.common.world.LevelChunkAuxiliaryLightManager.LIGHT_NBT_KEY).orElse(null);
+
             return new SerializableChunkData(
                 containerFactory,
                 chunkPos,
@@ -203,7 +235,9 @@
                 sectionData,
                 entities,
                 blockEntities,
-                structureData
+                structureData,
+                attachmentData,
+                auxLightData
             );
         }
     }
@@ -304,6 +341,10 @@
             }
         }
 
+        if (this.attachmentData != null) {
+            chunk.readAttachmentsFromNBT(level.registryAccess(), this.attachmentData);
+        }
+
         if (chunkType == ChunkType.LEVELCHUNK) {
             return new ImposterProtoChunk((LevelChunk)chunk, false);
         } else {
@@ -374,7 +415,7 @@
             Map<Heightmap.Types, long[]> heightmaps = new EnumMap<>(Heightmap.Types.class);
 
             for (Entry<Heightmap.Types, Heightmap> entry : chunk.getHeightmaps()) {
-                if (chunk.getPersistedStatus().heightmapsAfter().contains(entry.getKey())) {
+                if (chunk.getPersistedStatus().getChunkSaveHeightmaps().contains(entry.getKey())) {
                     long[] data = entry.getValue().getRawData();
                     heightmaps.put(entry.getKey(), (long[])data.clone());
                 }
@@ -387,6 +428,18 @@
             CompoundTag structureData = packStructureData(
                 StructurePieceSerializationContext.fromLevel(level), pos, chunk.getAllStarts(), chunk.getAllReferences()
             );
+
+            CompoundTag attachmentData = null;
+            try {
+                attachmentData = chunk.writeAttachmentsToNBT(level.registryAccess());
+            } catch (Exception exception) {
+                LOGGER.error("Failed to write chunk attachments. An attachment has likely thrown an exception trying to write state. It will not persist. Report this to the mod author", exception);
+            }
+            ListTag auxLightData = null;
+            if (chunk instanceof LevelChunk levelChunk) {
+                auxLightData = levelChunk.getAuxLightManager(pos).serializeNBT();
+            }
+
             return new SerializableChunkData(
                 level.palettedContainerFactory(),
                 pos,
@@ -405,7 +458,9 @@
                 sectionData,
                 entities,
                 blockEntities,
-                structureData
+                structureData,
+                attachmentData,
+                auxLightData
             );
         }
     }
