--- a/net/minecraft/client/renderer/block/dispatch/BlockStateModel.java
+++ b/net/minecraft/client/renderer/block/dispatch/BlockStateModel.java
@@ -22,13 +22,21 @@
 
 @OnlyIn(Dist.CLIENT)
 public interface BlockStateModel extends net.neoforged.neoforge.client.extensions.BlockStateModelExtension {
+    /// @deprecated Neo: Use [#collectParts(net.minecraft.client.renderer.block.BlockAndTintGetter, net.minecraft.core.BlockPos, net.minecraft.world.level.block.state.BlockState, RandomSource, List)].
+    @Deprecated
     void collectParts(RandomSource random, List<BlockStateModelPart> output);
 
+    /// @deprecated Neo: Use [#particleMaterial(net.minecraft.client.renderer.block.BlockAndTintGetter, net.minecraft.core.BlockPos, net.minecraft.world.level.block.state.BlockState)].
+    @Deprecated
     Material.Baked particleMaterial();
 
+    /// @deprecated Neo: Use [#materialFlags(net.minecraft.client.renderer.block.BlockAndTintGetter, net.minecraft.core.BlockPos, BlockState)] instead
+    @Deprecated
     @BakedQuad.MaterialFlags
     int materialFlags();
 
+    /// @deprecated Neo: Use [#hasMaterialFlag(net.minecraft.client.renderer.block.BlockAndTintGetter, net.minecraft.core.BlockPos, BlockState, int)] instead
+    @Deprecated
     default boolean hasMaterialFlag(@BakedQuad.MaterialFlags int flag) {
         return (this.materialFlags() & flag) != 0;
     }
