--- a/net/minecraft/client/renderer/texture/atlas/SpriteSourceList.java
+++ b/net/minecraft/client/renderer/texture/atlas/SpriteSourceList.java
@@ -35,7 +35,15 @@
         this.sources = sources;
     }
 
+    /**
+     * @deprecated Neo: use {@link #list(ResourceManager, java.util.Set)} instead
+     */
+    @Deprecated
     public List<SpriteSource.Loader> list(ResourceManager resourceManager) {
+        return this.list(resourceManager, java.util.Set.of());
+    }
+
+    public List<SpriteSource.Loader> list(ResourceManager resourceManager, java.util.Set<net.minecraft.server.packs.metadata.MetadataSectionType<?>> additionalMetadata) {
         final Map<Identifier, SpriteSource.DiscardableLoader> sprites = new HashMap<>();
         SpriteSource.Output output = new SpriteSource.Output() {
             {
