--- a/net/minecraft/client/gui/components/BossHealthOverlay.java
+++ b/net/minecraft/client/gui/components/BossHealthOverlay.java
@@ -75,6 +_,8 @@
                 int y = yo - 9;
                 graphics.text(this.minecraft.font, msg, x, y, -1);
                 yOffset += 10 + 9;
+                }
+                yOffset += customizeEvent.getIncrement();
                 if (yOffset >= graphics.guiHeight() / 3) {
                     break;
                 }
