--- a/net/minecraft/data/recipes/RecipeProvider.java
+++ b/net/minecraft/data/recipes/RecipeProvider.java
@@ -877,18 +880,14 @@
                         final Set<ResourceKey<Recipe<?>>> allRecipes = Sets.newHashSet();
                         final List<CompletableFuture<?>> tasks = new ArrayList<>();
                         RecipeOutput recipeOutput = new RecipeOutput() {
-                            {
-                                Objects.requireNonNull(Runner.this);
-                            }
-
                             @Override
-                            public void accept(ResourceKey<Recipe<?>> id, Recipe<?> recipe, @Nullable AdvancementHolder advancementHolder) {
+                            public void accept(ResourceKey<Recipe<?>> id, Recipe<?> recipe, @Nullable AdvancementHolder advancementHolder, net.neoforged.neoforge.common.conditions.ICondition... conditions) {
                                 if (!allRecipes.add(id)) {
                                     throw new IllegalStateException("Duplicate recipe " + id.identifier());
                                 } else {
-                                    this.saveRecipe(id, recipe);
+                                    this.saveRecipe(id, recipe, conditions);
                                     if (advancementHolder != null) {
-                                        this.saveAdvancement(advancementHolder);
+                                        this.saveAdvancement(advancementHolder, conditions);
                                     }
                                 }
                             }
