--- a/net/minecraft/client/multiplayer/ClientCommonPacketListenerImpl.java
+++ b/net/minecraft/client/multiplayer/ClientCommonPacketListenerImpl.java
@@ -504,5 +539,15 @@
         @OnlyIn(Dist.CLIENT)
         private record PendingRequest(UUID id, URL url, String hash) {
         }
+    }
+
+    @Override
+    public Connection getConnection() {
+        return connection;
+    }
+
+    @Override
+    public net.minecraft.network.PacketProcessor getPacketProcessor() {
+        return this.minecraft.packetProcessor();
     }
 }
