--- a/net/minecraft/client/multiplayer/ClientLevel.java
+++ b/net/minecraft/client/multiplayer/ClientLevel.java
@@ -161,6 +162,7 @@
     private final EnvironmentAttributeSystem environmentAttributes;
     private final int seaLevel;
     private static final Set<Item> MARKER_PARTICLE_ITEMS = Set.of(Items.BARRIER, Items.LIGHT);
+    private final net.neoforged.neoforge.model.data.ModelDataManager modelDataManager = new net.neoforged.neoforge.model.data.ModelDataManager(this);
 
     public void handleBlockChangedAck(int sequence) {
         if (SharedConstants.DEBUG_BLOCK_BREAK) {
