--- a/net/minecraft/client/multiplayer/ClientConfigurationPacketListenerImpl.java
+++ b/net/minecraft/client/multiplayer/ClientConfigurationPacketListenerImpl.java
@@ -182,13 +190,21 @@
                         this.customReportDetails,
                         this.serverLinks(),
                         this.seenPlayers,
-                        this.seenInsecureChatWarning
+                        this.seenInsecureChatWarning,
+                        this.connectionType
                     )
                 )
             );
+        net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.TagsUpdatedEvent.ClientPacketReceived(registries, this.connection.isMemoryConnection()));
+        // Packets can only be sent after the outbound protocol is set up again
+        if (!this.initializedConnection && this.connectionType.isOther()) {
+            // Neo: Fallback detection for servers with a delayed brand payload (BungeeCord)
+            net.neoforged.neoforge.client.network.registration.ClientNetworkRegistry.initializeOtherConnection(this);
+        }
+        net.neoforged.neoforge.network.registration.NetworkRegistry.onConfigurationFinished(this);
         this.connection.send(ServerboundFinishConfigurationPacket.INSTANCE);
         this.connection
-            .setupOutboundProtocol(GameProtocols.SERVERBOUND_TEMPLATE.bind(RegistryFriendlyByteBuf.decorator(registries), new GameProtocols.Context() {
+            .setupOutboundProtocol(GameProtocols.SERVERBOUND_TEMPLATE.bind(RegistryFriendlyByteBuf.decorator(registries, connectionType), new GameProtocols.Context() {
                 {
                     Objects.requireNonNull(ClientConfigurationPacketListenerImpl.this);
                 }
