--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -159,6 +159,7 @@
 import org.slf4j.Logger;
 
 public abstract class Entity
+    extends net.neoforged.neoforge.attachment.AttachmentHolder
     implements Nameable,
     EntityAccess,
     ScoreHolder,
@@ -205,6 +206,7 @@
     private static final int MAX_BLOCK_ITERATIONS_ALONG_TRAVEL_PER_TICK = 16;
     private static final double MAX_MOVEMENT_RESETTING_TRACE_DISTANCE = 8.0;
     private static double viewScale = 1.0;
+    @Deprecated // Neo: Use the getter to allow overriding in mods
     private final EntityType<?> type;
     private boolean requiresPrecisePosition;
     private int id = ENTITY_COUNTER.incrementAndGet();
@@ -996,9 +1001,7 @@
                 return getOnPos;
             } else {
                 BlockState belowState = this.level().getBlockState(getOnPos);
-                return (!(offset <= 0.5) || !belowState.is(BlockTags.FENCES))
-                        && !belowState.is(BlockTags.WALLS)
-                        && !(belowState.getBlock() instanceof FenceGateBlock)
+                return (!(offset <= 0.5) || !belowState.collisionExtendsVertically(this.level(), getOnPos, this))
                     ? getOnPos.atY(Mth.floor(this.position.y - offset))
                     : getOnPos;
             }
@@ -2125,6 +2136,8 @@
         } else {
             ItemEntity entity = new ItemEntity(level, this.getX() + offset.x, this.getY() + offset.y, this.getZ() + offset.z, itemStack);
             entity.setDefaultPickUpDelay();
+            if (captureDrops() != null) captureDrops().add(entity);
+            else
             level.addFreshEntity(entity);
             return entity;
         }
@@ -2355,6 +2372,7 @@
                 }
             }
 
+        if (!net.neoforged.neoforge.event.EventHooks.canMountEntity(this, entityToRide, true)) return false;
             if (force || this.canRide(entityToRide) && entityToRide.canAddPassenger(this)) {
                 if (this.isPassenger()) {
                     this.stopRiding();
