--- a/net/minecraft/world/entity/ExperienceOrb.java
+++ b/net/minecraft/world/entity/ExperienceOrb.java
@@ -136,7 +136,8 @@
             this.applyEffectsFromBlocks();
             float friction = 0.98F;
             if (this.onGround()) {
-                friction = this.level().getBlockState(this.getBlockPosBelowThatAffectsMyMovement()).getBlock().getFriction() * 0.98F;
+                BlockPos pos = this.getBlockPosBelowThatAffectsMyMovement();
+                friction = this.level().getBlockState(pos).getFriction(this.level(), pos, this) * 0.98F;
             }
 
             this.setDeltaMovement(this.getDeltaMovement().scale(friction));
