--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -1212,6 +1247,7 @@
             } else {
                 params.type().convert(this, newMob, params);
                 afterConversion.finalizeConversion(newMob);
+                if (!preventConversionSpawns)
                 if (this.level() instanceof ServerLevel serverLevel) {
                     serverLevel.addFreshEntity(newMob);
                 }
