--- a/net/minecraft/client/Minecraft.java
+++ b/net/minecraft/client/Minecraft.java
@@ -404,7 +_,6 @@
             }
         }, Util.nonCriticalIoPool());
         LOGGER.info("Setting user: {}", this.user.getName());
-        LOGGER.debug("(Session ID is {})", this.user.getSessionId());
         this.demo = gameConfig.game.demo;
         this.allowsMultiplayer = !gameConfig.game.disableMultiplayer;
         this.allowsChat = !gameConfig.game.disableChat;
@@ -740,6 +_,7 @@
     }
 
     private void onResourceLoadFinished(@Nullable GameLoadCookie loadCookie) {
+        net.neoforged.neoforge.client.ClientHooks.fireResourceLoadFinishedEvent(!this.gameLoadFinished);
         if (!this.gameLoadFinished) {
             this.gameLoadFinished = true;
             this.onGameLoadFinished(loadCookie);
@@ -775,7 +_,7 @@
     private String createTitle() {
         StringBuilder builder = new StringBuilder("Minecraft");
         if (checkModStatus().shouldReportAsModified()) {
-            builder.append("*");
+            builder.append(' ').append(net.neoforged.neoforge.internal.BrandingControl.BRANDING_NAME).append('*');
         }
 
         builder.append(" ");
@@ -815,7 +_,7 @@
     }
 
     private void rollbackResourcePacks(Throwable t, @Nullable GameLoadCookie loadCookie) {
-        if (this.resourcePackRepository.getSelectedIds().size() > 1) {
+        if (this.resourcePackRepository.getSelectedPacks().stream().anyMatch(e -> !e.isRequired())) { //Forge: This caused infinite loop if any resource packs are forced. Such as mod resources. So check if we can disable any.
             this.clearResourcePacksOnError(t, null, loadCookie);
         } else {
             Util.throwAsRuntime(t);
@@ -871,6 +_,8 @@
         DiscontinuousFrame tickFrame = TracyClient.createDiscontinuousFrame("Client Tick");
 
         try {
+            net.neoforged.neoforge.client.ClientLifecycleHooks.handleClientStarted(this);
+
             boolean oomRecovery = false;
 
             while (this.running) {
@@ -932,7 +_,7 @@
         CrashReport finalReport = this.fillReport(partialReport);
         int exitCode = saveReportAndShutdownSoundManager(this, this.gameDirectory, finalReport, -1);
         this.emergencySave();
-        System.exit(exitCode);
+        net.neoforged.neoforge.server.ServerLifecycleHooks.handleExit(exitCode);
     }
 
     public static void saveReport(File gameDirectory, CrashReport crash) {
@@ -1395,8 +_,10 @@
     public void resizeGui() {
         int guiScale = this.window.calculateScale(this.options.guiScale().get(), this.isEnforceUnicode());
         this.window.setGuiScale(guiScale);
+        net.neoforged.neoforge.client.ClientHooks.fireWindowResize(this.window);
         if (this.gui.screen() != null) {
             this.gui.screen().resize(this.window.getGuiScaledWidth(), this.window.getGuiScaledHeight());
+            net.neoforged.neoforge.client.ClientHooks.resizeGuiLayers(this.window.getGuiScaledWidth(), this.window.getGuiScaledHeight());
         }
 
         this.mouseHandler.setIgnoreFirstMove();
@@ -1548,6 +_,7 @@
     }
 
     public void stop() {
+        if (this.isRunning()) net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.GameShuttingDownEvent());
         this.running = false;
     }
 
@@ -1573,8 +_,16 @@
                     BlockPos pos = blockHit.getBlockPos();
                     if (!this.level.getBlockState(pos).isAir()) {
                         Direction direction = blockHit.getDirection();
-                        if (this.gameMode.continueDestroyBlock(pos, direction)) {
-                            this.level.addBreakingBlockEffect(pos, direction);
+                        var inputEvent = net.neoforged.neoforge.client.ClientHooks.onClickInput(0, this.options.keyAttack, InteractionHand.MAIN_HAND);
+                        if (inputEvent.isCanceled()) {
+                            if (inputEvent.shouldSwingHand()) {
+                                this.level.addBreakingBlockEffect(pos, direction, blockHit);
+                                this.player.swing(InteractionHand.MAIN_HAND);
+                            }
+                            return;
+                        }
+                        if (this.gameMode.continueDestroyBlock(pos, direction) && inputEvent.shouldSwingHand()) {
+                            this.level.addBreakingBlockEffect(pos, direction, blockHit);
                             this.player.swing(InteractionHand.MAIN_HAND);
                         }
                     }
@@ -1668,6 +_,11 @@
                 }
 
                 for (InteractionHand hand : InteractionHand.values()) {
+                    var inputEvent = net.neoforged.neoforge.client.ClientHooks.onClickInput(1, this.options.keyUse, hand);
+                    if (inputEvent.isCanceled()) {
+                        if (inputEvent.shouldSwingHand()) this.player.swing(hand);
+                        return;
+                    }
                     ItemStack heldItem = this.player.getItemInHand(hand);
                     if (!heldItem.isItemEnabled(this.level.enabledFeatures())) {
                         return;
@@ -1684,7 +_,7 @@
 
                                 if (this.player.isWithinEntityInteractionRange(entity, 0.0)
                                     && this.gameMode.interact(this.player, entity, entityHit, hand) instanceof InteractionResult.Success success) {
-                                    if (success.swingSource() == InteractionResult.SwingSource.CLIENT) {
+                                    if (success.swingSource() == InteractionResult.SwingSource.CLIENT && inputEvent.shouldSwingHand()) {
                                         this.player.swing(hand);
                                     }
 
@@ -1696,7 +_,7 @@
                                 int oldCount = heldItem.getCount();
                                 InteractionResult useResult = this.gameMode.useItemOn(this.player, hand, blockHit);
                                 if (useResult instanceof InteractionResult.Success success) {
-                                    if (success.swingSource() == InteractionResult.SwingSource.CLIENT) {
+                                    if (success.swingSource() == InteractionResult.SwingSource.CLIENT && inputEvent.shouldSwingHand()) {
                                         this.player.swing(hand);
                                         if (!heldItem.isEmpty() && (heldItem.getCount() != oldCount || this.player.hasInfiniteMaterials())) {
                                             this.gameRenderer.itemInHandRenderer.itemUsed(hand);
@@ -1712,8 +_,11 @@
                         }
                     }
 
+                    if (heldItem.isEmpty() && (this.hitResult == null || this.hitResult.getType() == HitResult.Type.MISS))
+                        net.neoforged.neoforge.common.CommonHooks.onEmptyClick(this.player, hand);
+
                     if (!heldItem.isEmpty() && this.gameMode.useItem(this.player, hand) instanceof InteractionResult.Success success) {
-                        if (success.swingSource() == InteractionResult.SwingSource.CLIENT) {
+                        if (success.swingSource() == InteractionResult.SwingSource.CLIENT && inputEvent.shouldSwingHand()) {
                             this.player.swing(hand);
                         }
 
@@ -1731,6 +_,10 @@
 
     public void tick() {
         this.clientTickCount++;
+        if (this.gameLoadFinished) {
+            net.neoforged.neoforge.client.ClientHooks.fireClientTickPre();
+        }
+
         if (this.level != null && !this.pause) {
             this.level.tickRateManager().tick();
         }
@@ -1793,6 +_,7 @@
 
                 this.tutorial.tick();
 
+                net.neoforged.neoforge.event.EventHooks.fireLevelTickPre(this.level, () -> true);
                 try {
                     this.level.tick(() -> true);
                 } catch (Throwable t) {
@@ -1806,6 +_,7 @@
 
                     throw new ReportedException(report);
                 }
+                net.neoforged.neoforge.event.EventHooks.fireLevelTickPost(this.level, () -> true);
             }
 
             profiler.popPush("animateTick");
@@ -1830,6 +_,10 @@
         profiler.popPush("keyboard");
         this.keyboardHandler.tick();
         profiler.pop();
+
+        if (this.gameLoadFinished) {
+            net.neoforged.neoforge.client.ClientHooks.fireClientTickPost();
+        }
     }
 
     private boolean isLevelRunningNormally() {
@@ -2035,6 +_,7 @@
     }
 
     public void setLevel(ClientLevel level) {
+        if (this.level != null) net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.level.LevelEvent.Unload(this.level));
         this.level = level;
         this.updateLevelInEngines(level);
     }
@@ -2151,6 +_,7 @@
         try {
             this.setScreenAndShow(screen);
             this.gui.hud.onDisconnected();
+            if (this.level != null) net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.level.LevelEvent.Unload(this.level));
             this.level = null;
             this.updateLevelInEngines(null);
             this.player = null;
@@ -2283,6 +_,7 @@
 
     private void pickBlockOrEntity() {
         if (this.hitResult != null && this.hitResult.getType() != HitResult.Type.MISS) {
+            if (net.neoforged.neoforge.client.ClientHooks.onClickInput(2, this.options.keyPickItem, InteractionHand.MAIN_HAND).isCanceled()) return;
             boolean includeData = this.hasControlDown();
             switch (this.hitResult) {
                 case BlockHitResult blockHitResult:
@@ -2832,6 +_,10 @@
         }
 
         return defaultTickTargetMillis;
+    }
+
+    public BlockModelResolver getBlockModelResolver() {
+        return this.blockModelResolver;
     }
 
     public ItemModelResolver getItemModelResolver() {
