--- a/net/minecraft/client/Screenshot.java
+++ b/net/minecraft/client/Screenshot.java
@@ -51,6 +_,13 @@
                     file = new File(picDir, forceName);
                 }
 
+                net.neoforged.neoforge.client.event.ScreenshotEvent event = net.neoforged.neoforge.client.ClientHooks.onScreenshot(image, file);
+                if (event.isCanceled()) {
+                    callback.accept(event.getCancelMessage());
+                    return;
+                }
+                final File screenshotFile = event.getScreenshotFile();
+
                 Util.ioPool()
                     .execute(
                         () -> {
