--- a/com/mojang/blaze3d/pipeline/RenderPipeline.java
+++ b/com/mojang/blaze3d/pipeline/RenderPipeline.java
@@ -130,6 +_,45 @@
         return this.depthStencilState != null;
     }
 
+    public Optional<net.neoforged.neoforge.client.stencil.StencilTest> getStencilTest() {
+        return stencilTest;
+    }
+
+    /**
+     * Neo: Create a {@link RenderPipeline.Builder} from this {@link RenderPipeline} to adjust its configuration and
+     * build a new, modified {@link RenderPipeline} from it
+     */
+    public RenderPipeline.Builder toBuilder() {
+        RenderPipeline.Builder builder = new RenderPipeline.Builder();
+        builder.location = Optional.of(this.location);
+        builder.fragmentShader = Optional.of(this.fragmentShader);
+        builder.vertexShader = Optional.of(this.vertexShader);
+        if (!this.shaderDefines.isEmpty()) {
+            ShaderDefines.Builder defBuilder = ShaderDefines.builder();
+            for (Entry<String, String> entry : this.shaderDefines.values().entrySet()) {
+                defBuilder.define(entry.getKey(), entry.getValue());
+            }
+            for (String flag : this.shaderDefines.flags()) {
+                defBuilder.define(flag);
+            }
+            builder.definesBuilder = Optional.of(defBuilder);
+        }
+        if (!this.samplers.isEmpty()) {
+            builder.samplers = Optional.of(new ArrayList<>(this.samplers));
+        }
+        if (!this.uniforms.isEmpty()) {
+            builder.uniforms = Optional.of(new ArrayList<>(this.uniforms));
+        }
+        builder.depthStencilState = Optional.ofNullable(this.depthStencilState);
+        builder.polygonMode = Optional.of(this.polygonMode);
+        builder.cull = Optional.of(this.cull);
+        builder.colorTargetState = Optional.of(this.colorTargetState);
+        builder.vertexFormat = Optional.of(this.vertexFormat);
+        builder.vertexFormatMode = Optional.of(this.vertexFormatMode);
+        builder.stencilTest = this.stencilTest;
+        return builder;
+    }
+
     public static RenderPipeline.Builder builder(RenderPipeline.Snippet... snippets) {
         RenderPipeline.Builder builder = new RenderPipeline.Builder();
 
@@ -270,6 +_,16 @@
             return this;
         }
 
+        public RenderPipeline.Builder withStencilTest(net.neoforged.neoforge.client.stencil.StencilTest stencilTest) {
+            this.stencilTest = Optional.of(stencilTest);
+            return this;
+        }
+
+        public RenderPipeline.Builder withoutStencilTest(){
+            this.stencilTest = Optional.empty();
+            return this;
+        }
+
         private void withSnippet(RenderPipeline.Snippet snippet) {
             if (snippet.vertexShader.isPresent()) {
                 this.vertexShader = snippet.vertexShader;
@@ -331,6 +_,10 @@
 
             if (snippet.polygonMode.isPresent()) {
                 this.polygonMode = snippet.polygonMode;
+            }
+
+            if (snippet.stencilTest.isPresent()) {
+                this.stencilTest = snippet.stencilTest;
             }
         }
 
