--- a/com/mojang/blaze3d/pipeline/RenderTarget.java
+++ b/com/mojang/blaze3d/pipeline/RenderTarget.java
@@ -83,7 +_,8 @@
             this.width = width;
             this.height = height;
             if (this.useDepth) {
-                this.depthTexture = device.createTexture(() -> this.label + " / Depth", 15, GpuFormat.D32_FLOAT, width, height, 1, 1);
+                var format = this.useStencil ? net.neoforged.neoforge.client.ClientHooks.getStencilFormat() : TextureFormat.DEPTH32;
+                this.depthTexture = device.createTexture(() -> this.label + " / Depth", 15, format, width, height, 1, 1);
                 this.depthTextureView = device.createTextureView(this.depthTexture);
             }
 
