--- a/net/minecraft/world/damagesource/DeathMessageType.java
+++ b/net/minecraft/world/damagesource/DeathMessageType.java
@@ -3,6 +3,8 @@
 import com.mojang.serialization.Codec;
 import net.minecraft.util.StringRepresentable;
 
+@net.neoforged.fml.common.asm.enumextension.NamedEnum
+@net.neoforged.fml.common.asm.enumextension.NetworkedEnum(net.neoforged.fml.common.asm.enumextension.NetworkedEnum.NetworkCheck.CLIENTBOUND)
 public enum DeathMessageType implements StringRepresentable, net.neoforged.fml.common.asm.enumextension.IExtensibleEnum {
     DEFAULT("default"),
     FALL_VARIANTS("fall_variants"),
