--- a/net/minecraft/world/level/gamerules/GameRuleType.java
+++ b/net/minecraft/world/level/gamerules/GameRuleType.java
@@ -18,4 +_,8 @@
     public String getSerializedName() {
         return this.name;
     }
+
+    public static net.neoforged.fml.common.asm.enumextension.ExtensionInfo getExtensionInfo() {
+        return net.neoforged.fml.common.asm.enumextension.ExtensionInfo.nonExtended(GameRuleType.class);
+    }
 }
