--- a/net/minecraft/world/level/block/piston/PistonStructureResolver.java
+++ b/net/minecraft/world/level/block/piston/PistonStructureResolver.java
@@ -50,7 +50,7 @@
         } else {
             for (int i = 0; i < this.toPush.size(); i++) {
                 BlockPos pos = this.toPush.get(i);
-                if (isSticky(this.level.getBlockState(pos)) && !this.addBranchingBlocks(pos)) {
+                if (this.level.getBlockState(pos).isStickyBlock() && !this.addBranchingBlocks(pos)) {
                     return false;
                 }
             }
@@ -86,12 +74,13 @@
             if (blockCount + this.toPush.size() > 12) {
                 return false;
             } else {
-                while (isSticky(nextState)) {
+                BlockState oldState;
+                while(nextState.isStickyBlock()) {
                     BlockPos pos = start.relative(this.pushDirection.getOpposite(), blockCount);
-                    BlockState previousState = nextState;
+                    oldState = nextState;
                     nextState = this.level.getBlockState(pos);
                     if (nextState.isAir()
-                        || !canStickToEachOther(previousState, nextState)
+                        || !(oldState.canStickTo(nextState) && oldState.canStickTo(oldState))
                         || !PistonBaseBlock.isPushable(nextState, this.level, pos, this.pushDirection, false, this.pushDirection.getOpposite())
                         || pos.equals(this.pistonPos)) {
                         break;
@@ -119,7 +108,7 @@
 
                         for (int j = 0; j <= collisionPos + blocksAdded; j++) {
                             BlockPos blockPos = this.toPush.get(j);
-                            if (isSticky(this.level.getBlockState(blockPos)) && !this.addBranchingBlocks(blockPos)) {
+                            if (this.level.getBlockState(blockPos).isStickyBlock() && !this.addBranchingBlocks(blockPos)) {
                                 return false;
                             }
                         }
