--- a/net/minecraft/world/inventory/GrindstoneMenu.java
+++ b/net/minecraft/world/inventory/GrindstoneMenu.java
@@ -37,6 +_,7 @@
         }
     };
     private final ContainerLevelAccess access;
+    private int xp = -1;
 
     public GrindstoneMenu(int containerId, Inventory inventory) {
         this(containerId, inventory, ContainerLevelAccess.NULL);
@@ -65,6 +_,7 @@
 
             @Override
             public void onTake(Player player, ItemStack carried) {
+                if (net.neoforged.neoforge.common.CommonHooks.onGrindstoneTake(GrindstoneMenu.this.repairSlots, access, player, this::getExperienceAmount)) return;
                 access.execute((level, pos) -> {
                     if (level instanceof ServerLevel serverLevel) {
                         ExperienceOrb.award(serverLevel, Vec3.atCenterOf(pos), this.getExperienceAmount(level));
@@ -77,6 +_,7 @@
             }
 
             private int getExperienceAmount(Level level) {
+                if (xp > -1) return xp;
                 int amount = 0;
                 amount += this.getExperienceFromItem(GrindstoneMenu.this.repairSlots.getItem(0));
                 amount += this.getExperienceFromItem(GrindstoneMenu.this.repairSlots.getItem(1));
@@ -115,6 +_,8 @@
     }
 
     private void createResult() {
+        this.xp = net.neoforged.neoforge.common.CommonHooks.onGrindstoneChange(this.repairSlots.getItem(0), this.repairSlots.getItem(1), this.resultSlots, -1);
+        if (this.xp == Integer.MIN_VALUE)
         this.resultSlots.setItem(0, this.computeResult(this.repairSlots.getItem(0), this.repairSlots.getItem(1)));
         this.broadcastChanges();
     }
