--- a/net/minecraft/world/entity/vehicle/minecart/NewMinecartBehavior.java
+++ b/net/minecraft/world/entity/vehicle/minecart/NewMinecartBehavior.java
@@ -387,7 +387,7 @@
     }
 
     private Vec3 calculateBoostTrackSpeed(Vec3 deltaMovement, BlockPos pos, BlockState state) {
-        if (state.is(Blocks.POWERED_RAIL) && state.getValue(PoweredRailBlock.POWERED)) {
+        if (state.getBlock() instanceof PoweredRailBlock poweredRail && !poweredRail.isActivatorRail() && state.getValue(PoweredRailBlock.POWERED)) {
             if (deltaMovement.length() > 0.01) {
                 return deltaMovement.normalize().scale(deltaMovement.length() + 0.06);
             } else {
@@ -436,10 +436,11 @@
                 }
 
                 this.minecart.move(MoverType.SELF, newPosition.subtract(oldPosition));
-                BlockState newBlockState = this.level().getBlockState(BlockPos.containing(newPosition));
+                BlockPos railPos = BlockPos.containing(newPosition);
+                BlockState newBlockState = this.level().getBlockState(railPos);
                 if (inHill) {
                     if (BaseRailBlock.isRail(newBlockState)) {
-                        RailShape newRailShape = newBlockState.getValue(((BaseRailBlock)newBlockState.getBlock()).getShapeProperty());
+                        RailShape newRailShape = ((BaseRailBlock)newBlockState.getBlock()).getRailDirection(newBlockState, this.level(), railPos, this.minecart);;
                         if (this.restAtVShape(shape, newRailShape)) {
                             return 0.0;
                         }
