--- a/net/minecraft/resources/RegistryLoadTask.java
+++ b/net/minecraft/resources/RegistryLoadTask.java
@@ -105,7 +117,15 @@
                 Either var6;
                 try (Reader reader = thunk.openAsReader()) {
                     JsonElement json = StrictJsonParser.parse(reader);
-                    var6 = Either.left(elementDecoder.parse(ops, json).getOrThrow());
+
+                    // Neo: Allow a condition to skip the entire element, in which case we return a special marker error to skip it in the loader. Look for uses of SKIPPED_ELEMENT_MARKER
+                    var conditionalDecoder = net.neoforged.neoforge.common.conditions.ConditionalOps.createConditionalCodec(net.neoforged.neoforge.common.util.NeoForgeExtraCodecs.decodeOnly(elementDecoder));
+                    var candidate = conditionalDecoder.parse(ops, json).getOrThrow();
+                    if (candidate.isPresent()) {
+                        var6 = Either.left(candidate.get());
+                    } else {
+                        var6 = Either.right(SKIPPED_ELEMENT_MARKER);
+                    }
                 }
 
                 return var6;
