--- a/net/minecraft/client/gui/render/pip/PictureInPictureRenderer.java
+++ b/net/minecraft/client/gui/render/pip/PictureInPictureRenderer.java
@@ -96,7 +_,9 @@
         if (this.texture == null) {
             this.texture = device.createTexture(() -> "UI " + this.getTextureLabel() + " texture", 13, GpuFormat.RGBA8_UNORM, width, height, 1, 1);
             this.textureView = device.createTextureView(this.texture);
-            this.depthTexture = device.createTexture(() -> "UI " + this.getTextureLabel() + " depth texture", 9, GpuFormat.D32_FLOAT, width, height, 1, 1);
+            // Neo: copy stencil setting from main target
+            TextureFormat depthFormat = net.minecraft.client.Minecraft.getInstance().getMainRenderTarget().getDepthTexture().getFormat();
+            this.depthTexture = device.createTexture(() -> "UI " + this.getTextureLabel() + " depth texture", 9, depthFormat, width, height, 1, 1);
             this.depthTextureView = device.createTextureView(this.depthTexture);
         }
 
