--- a/net/minecraft/core/MappedRegistry.java
+++ b/net/minecraft/core/MappedRegistry.java
@@ -30,7 +_,7 @@
 import net.minecraft.util.Util;
 import org.jspecify.annotations.Nullable;
 
-public class MappedRegistry<T> implements WritableRegistry<T> {
+public class MappedRegistry<T> extends net.neoforged.neoforge.registries.BaseMappedRegistry<T> implements WritableRegistry<T> {
     private final ResourceKey<? extends Registry<T>> key;
     private final ObjectList<Holder.Reference<T>> byId = new ObjectArrayList<>(256);
     private final Reference2IntMap<T> toId = Util.make(new Reference2IntOpenHashMap<>(), t -> t.defaultReturnValue(-1));
@@ -138,7 +_,7 @@
 
     @Override
     public @Nullable T getValue(@Nullable ResourceKey<T> key) {
-        return getValueFromNullable(this.byKey.get(key));
+        return getValueFromNullable(this.byKey.get(resolve(key)));
     }
 
     @Override
@@ -153,12 +_,12 @@
 
     @Override
     public Optional<Holder.Reference<T>> get(Identifier id) {
-        return Optional.ofNullable(this.byLocation.get(id));
+        return Optional.ofNullable(this.byLocation.get(resolve(id)));
     }
 
     @Override
     public Optional<Holder.Reference<T>> get(ResourceKey<T> id) {
-        return Optional.ofNullable(this.byKey.get(id));
+        return Optional.ofNullable(this.byKey.get(resolve(id)));
     }
 
     @Override
@@ -205,7 +_,7 @@
 
     @Override
     public @Nullable T getValue(@Nullable Identifier key) {
-        Holder.Reference<T> result = this.byLocation.get(key);
+        Holder.Reference<T> result = this.byLocation.get(key != null ? resolve(key) : null);
         return getValueFromNullable(result);
     }
 
