--- a/net/minecraft/world/entity/monster/piglin/AbstractPiglin.java
+++ b/net/minecraft/world/entity/monster/piglin/AbstractPiglin.java
@@ -110,7 +_,10 @@
         this.convertTo(
             EntityTypes.ZOMBIFIED_PIGLIN,
             ConversionParams.single(this, true, true),
-            zombified -> zombified.addEffect(new MobEffectInstance(MobEffects.NAUSEA, 200, 0))
+            zombified -> {
+                zombified.addEffect(new MobEffectInstance(MobEffects.NAUSEA, 200, 0));
+                net.neoforged.neoforge.event.EventHooks.onLivingConvert(this, zombified);
+            }
         );
     }
 
