--- a/net/minecraft/client/renderer/state/level/LevelRenderState.java
+++ b/net/minecraft/client/renderer/state/level/LevelRenderState.java
@@ -9,7 +_,7 @@
 import org.jspecify.annotations.Nullable;
 
 @OnlyIn(Dist.CLIENT)
-public class LevelRenderState {
+public class LevelRenderState extends net.neoforged.neoforge.client.renderstate.BaseRenderState {
     public CameraRenderState cameraRenderState = new CameraRenderState();
     public final List<SectionUpdateRenderState> sectionUpdateRenderStates = new ArrayList<>();
     public final List<EntityRenderState> entityRenderStates = new ArrayList<>();
@@ -43,5 +_,9 @@
         this.particlesRenderState.reset();
         this.chunkLoadingRenderState.reset();
         this.gameTime = 0L;
+        this.customSkyboxRenderer = null;
+        this.customCloudsRenderer = null;
+        this.customWeatherEffectRenderer = null;
+        this.resetRenderData();
     }
 }
