--- a/com/mojang/blaze3d/systems/RenderPass.java
+++ b/com/mojang/blaze3d/systems/RenderPass.java
@@ -116,6 +_,10 @@
         this.backend.setUniform(name, value);
     }
 
+    public void setViewport(int x, int y, int width, int height) {
+        this.backend.setViewport(x, y, width, height);
+    }
+
     public void enableScissor(int x, int y, int width, int height) {
         if (width > 0 && height > 0) {
             if (x >= this.renderArea.x()
