--- a/net/minecraft/client/sounds/SoundEngine.java
+++ b/net/minecraft/client/sounds/SoundEngine.java
@@ -82,6 +_,7 @@
         this.options = options;
         this.soundBuffers = new SoundBufferLibrary(resourceProvider);
         this.lastSeenDevices = this.deviceTracker.currentDevices();
+        net.neoforged.fml.ModLoader.postEvent(new net.neoforged.neoforge.client.event.sound.SoundEngineLoadEvent(this));
     }
 
     public void reload() {
@@ -99,6 +_,7 @@
 
         this.destroy();
         this.loadLibrary();
+        net.neoforged.fml.ModLoader.postEvent(new net.neoforged.neoforge.client.event.sound.SoundEngineLoadEvent(this));
     }
 
     private synchronized void loadLibrary() {
