--- a/net/minecraft/world/entity/item/ItemEntity.java
+++ b/net/minecraft/world/entity/item/ItemEntity.java
@@ -142,7 +149,8 @@
                 this.applyEffectsFromBlocks();
                 float friction = 0.98F;
                 if (this.onGround()) {
-                    friction = this.level().getBlockState(this.getBlockPosBelowThatAffectsMyMovement()).getBlock().getFriction() * 0.98F;
+                    BlockPos groundPos = getBlockPosBelowThatAffectsMyMovement();
+                    friction = this.level().getBlockState(groundPos).getFriction(level(), groundPos, this) * 0.98F;
                 }
 
                 this.setDeltaMovement(this.getDeltaMovement().multiply(friction, 0.98, friction));
@@ -284,7 +300,7 @@
             this.health = (int)(this.health - damage);
             this.gameEvent(GameEvent.ENTITY_DAMAGE, source.getEntity());
             if (this.health <= 0) {
-                this.getItem().onDestroyed(this);
+                this.getItem().onDestroyed(this, source);
                 this.discard();
             }
 
