--- a/net/minecraft/world/level/block/WeatheringCopper.java
+++ b/net/minecraft/world/level/block/WeatheringCopper.java
@@ -63,7 +_,7 @@
     }
 
     static Optional<Block> getNext(Block block) {
-        return Optional.ofNullable(NEXT_BY_BLOCK.get().get(block));
+        return Optional.ofNullable(net.neoforged.neoforge.common.DataMapHooks.getNextOxidizedStage(block));
     }
 
     static BlockState getFirst(BlockState state) {
