--- a/net/minecraft/client/gui/screens/advancements/AdvancementTab.java
+++ b/net/minecraft/client/gui/screens/advancements/AdvancementTab.java
@@ -148,8 +158,8 @@
             return null;
         } else {
             for (AdvancementTabType type : AdvancementTabType.values()) {
-                if (index < type.getMax()) {
-                    return new AdvancementTab(minecraft, screen, type, index, root, display.get());
+                if ((index % AdvancementTabType.MAX_TABS) < type.getMax()) {
+                    return new AdvancementTab(minecraft, screen, type, index % AdvancementTabType.MAX_TABS, index / AdvancementTabType.MAX_TABS, root, display.get());
                 }
 
                 index -= type.getMax();
