--- a/net/minecraft/world/level/block/entity/HopperBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/HopperBlockEntity.java
@@ -139,10 +174,11 @@
     }
 
     private static boolean ejectItems(Level level, BlockPos blockPos, HopperBlockEntity self) {
-        Container container = getAttachedContainer(level, blockPos, self);
-        if (container == null) {
+        var containerOrHandler = getContainerOrHandlerAt(level, blockPos.relative(self.facing), self.facing.getOpposite());
+        if (containerOrHandler.isEmpty()) {
             return false;
-        } else {
+        } else if (containerOrHandler.container() != null) {
+            Container container = containerOrHandler.container();
             Direction direction = self.facing.getOpposite();
             if (isFullContainer(container, direction)) {
                 return false;
@@ -166,6 +202,8 @@
 
                 return false;
             }
+        } else {
+            return net.neoforged.neoforge.transfer.item.VanillaInventoryCodeHooks.insertHook(self, containerOrHandler.itemHandler());
         }
     }
 
