--- a/com/mojang/blaze3d/pipeline/RenderPipeline.java
+++ b/com/mojang/blaze3d/pipeline/RenderPipeline.java
@@ -32,7 +32,10 @@
     private final VertexFormat.Mode vertexFormatMode;
     private final int sortKey;
     private static int sortKeySeed;
+    private final Optional<net.neoforged.neoforge.client.stencil.StencilTest> stencilTest;
 
+    /** @deprecated Neo: use {@link #RenderPipeline(Identifier, Identifier, Identifier, ShaderDefines, List, List, ColorTargetState, DepthStencilState, PolygonMode, boolean, VertexFormat, VertexFormat.Mode, int, Optional)} instead */
+    @Deprecated
     protected RenderPipeline(
         Identifier location,
         Identifier vertexShader,
@@ -48,6 +51,25 @@
         VertexFormat.Mode vertexFormatMode,
         int sortKey
     ) {
+        this(location, vertexShader, fragmentShader, shaderDefines, samplers, uniforms, colorTargetState, depthStencilState, polygonMode, cull, vertexFormat, vertexFormatMode, sortKey, Optional.empty());
+    }
+
+    protected RenderPipeline(
+        Identifier location,
+        Identifier vertexShader,
+        Identifier fragmentShader,
+        ShaderDefines shaderDefines,
+        List<String> samplers,
+        List<RenderPipeline.UniformDescription> uniforms,
+        ColorTargetState colorTargetState,
+        @Nullable DepthStencilState depthStencilState,
+        PolygonMode polygonMode,
+        boolean cull,
+        VertexFormat vertexFormat,
+        VertexFormat.Mode vertexFormatMode,
+        int sortKey,
+        Optional<net.neoforged.neoforge.client.stencil.StencilTest> stencilTest
+    ) {
         this.location = location;
         this.vertexShader = vertexShader;
         this.fragmentShader = fragmentShader;
@@ -61,6 +83,7 @@
         this.vertexFormat = vertexFormat;
         this.vertexFormatMode = vertexFormatMode;
         this.sortKey = sortKey;
+        this.stencilTest = stencilTest;
     }
 
     public int getSortKey() {
@@ -153,6 +215,7 @@
         private Optional<ColorTargetState> colorTargetState = Optional.empty();
         private Optional<VertexFormat> vertexFormat = Optional.empty();
         private Optional<VertexFormat.Mode> vertexFormatMode = Optional.empty();
+        private Optional<net.neoforged.neoforge.client.stencil.StencilTest> stencilTest = Optional.empty();
 
         private Builder() {
         }
@@ -356,7 +433,8 @@
                 this.polygonMode,
                 this.cull,
                 this.vertexFormat,
-                this.vertexFormatMode
+                this.vertexFormatMode,
+                this.stencilTest
             );
         }
 
@@ -385,7 +463,8 @@
                     this.cull.orElse(true),
                     this.vertexFormat.get(),
                     this.vertexFormatMode.get(),
-                    nextPipelineSortKey++
+                    nextPipelineSortKey++,
+                    this.stencilTest
                 );
             }
         }
@@ -403,8 +482,26 @@
         Optional<PolygonMode> polygonMode,
         Optional<Boolean> cull,
         Optional<VertexFormat> vertexFormat,
-        Optional<VertexFormat.Mode> vertexFormatMode
+        Optional<VertexFormat.Mode> vertexFormatMode,
+        Optional<net.neoforged.neoforge.client.stencil.StencilTest> stencilTest
     ) {
+        /** @deprecated Neo: use {@link #Snippet(Optional, Optional, Optional, Optional, Optional, Optional, Optional, Optional, Optional, Optional, Optional, Optional)} instead */
+        @Deprecated
+        public Snippet(
+                Optional<Identifier> vertexShader,
+                Optional<Identifier> fragmentShader,
+                Optional<ShaderDefines> shaderDefines,
+                Optional<List<String>> samplers,
+                Optional<List<RenderPipeline.UniformDescription>> uniforms,
+                Optional<ColorTargetState> colorTargetState,
+                Optional<DepthStencilState> depthStencilState,
+                Optional<PolygonMode> polygonMode,
+                Optional<Boolean> cull,
+                Optional<VertexFormat> vertexFormat,
+                Optional<VertexFormat.Mode> vertexFormatMode
+        ) {
+            this(vertexShader, fragmentShader, shaderDefines, samplers, uniforms, colorTargetState, depthStencilState, polygonMode, cull, vertexFormat, vertexFormatMode, Optional.empty());
+        }
     }
 
     @OnlyIn(Dist.CLIENT)
