--- a/net/minecraft/client/gui/screens/inventory/CreativeModeInventoryScreen.java
+++ b/net/minecraft/client/gui/screens/inventory/CreativeModeInventoryScreen.java
@@ -152,9 +154,11 @@
             return false;
         } else {
             if (searchTrees != null) {
-                List<ItemStack> creativeSearchItems = List.copyOf(CreativeModeTabs.searchTab().getDisplayItems());
-                searchTrees.updateCreativeTooltips(holders, creativeSearchItems);
-                searchTrees.updateCreativeTags(creativeSearchItems);
+                CreativeModeTabs.allTabs().stream().filter(net.minecraft.world.item.CreativeModeTab::hasSearchBar).forEach(tab -> {
+                    List<ItemStack> list = List.copyOf(tab.getDisplayItems());
+                    searchTrees.updateCreativeTooltips(holders, list, net.neoforged.neoforge.client.CreativeModeTabSearchRegistry.getNameSearchKey(tab));
+                    searchTrees.updateCreativeTags(list, net.neoforged.neoforge.client.CreativeModeTabSearchRegistry.getTagSearchKey(tab));
+                });
             }
 
             return true;
@@ -377,7 +409,7 @@
     public boolean charTyped(CharacterEvent event) {
         if (this.ignoreTextInput) {
             return false;
-        } else if (selectedTab.getType() != CreativeModeTab.Type.SEARCH) {
+        } else if (!selectedTab.hasSearchBar()) {
             return false;
         } else {
             String oldContents = this.searchBox.getValue();
@@ -701,10 +742,10 @@
     public List<Component> getTooltipFromContainerItem(ItemStack itemStack) {
         boolean isCreativeSlot = this.hoveredSlot != null && this.hoveredSlot instanceof CreativeModeInventoryScreen.CustomCreativeSlot;
         boolean isSingleCategoryTab = selectedTab.getType() == CreativeModeTab.Type.CATEGORY;
-        boolean isSearchTab = selectedTab.getType() == CreativeModeTab.Type.SEARCH;
+        boolean isSearchTab = selectedTab.hasSearchBar();
         TooltipFlag.Default originalTooltipStyle = this.minecraft.options.advancedItemTooltips ? TooltipFlag.Default.ADVANCED : TooltipFlag.Default.NORMAL;
         TooltipFlag tooltipStyle = isCreativeSlot ? originalTooltipStyle.asCreative() : originalTooltipStyle;
-        List<Component> originalLines = itemStack.getTooltipLines(Item.TooltipContext.of(this.minecraft.level), this.minecraft.player, tooltipStyle);
+        List<Component> originalLines = itemStack.getTooltipLines(Item.TooltipContext.of(this.minecraft.level, this.minecraft.player), this.minecraft.player, net.neoforged.neoforge.client.ClientTooltipFlag.of(tooltipStyle));
         if (originalLines.isEmpty()) {
             return originalLines;
         } else if (isSingleCategoryTab && isCreativeSlot) {
@@ -722,7 +763,7 @@
             int i = 1;
 
             for (CreativeModeTab tab : CreativeModeTabs.tabs()) {
-                if (tab.getType() != CreativeModeTab.Type.SEARCH && tab.contains(itemStack)) {
+                if (!tab.hasSearchBar() && tab.contains(itemStack)) {
                     linesToDisplay.add(i++, tab.getDisplayName().copy().withStyle(ChatFormatting.BLUE));
                 }
             }
@@ -870,6 +913,14 @@
         }
     }
 
+    public net.neoforged.neoforge.client.gui.CreativeTabsScreenPage getCurrentPage() {
+        return currentPage;
+    }
+
+    public void setCurrentPage(net.neoforged.neoforge.client.gui.CreativeTabsScreenPage currentPage) {
+        this.currentPage = currentPage;
+    }
+
     @OnlyIn(Dist.CLIENT)
     private static class CustomCreativeSlot extends Slot {
         public CustomCreativeSlot(Container container, int slot, int x, int y) {
