--- a/net/minecraft/client/multiplayer/MultiPlayerGameMode.java
+++ b/net/minecraft/client/multiplayer/MultiPlayerGameMode.java
@@ -242,6 +_,7 @@
             }
 
             this.startPrediction(this.minecraft.level, sequence -> {
+                if (!net.neoforged.neoforge.common.CommonHooks.onLeftClickBlock(this.minecraft.player, pos, direction, ServerboundPlayerActionPacket.Action.START_DESTROY_BLOCK).isCanceled())
                 this.destroyBlock(pos);
                 return new ServerboundPlayerActionPacket(ServerboundPlayerActionPacket.Action.START_DESTROY_BLOCK, pos, direction, sequence);
             });
@@ -313,7 +_,7 @@
 
     private boolean sameDestroyTarget(BlockPos pos) {
         ItemStack selected = this.minecraft.player.getMainHandItem();
-        return pos.equals(this.destroyBlockPos) && ItemStack.isSameItemSameComponents(selected, this.destroyingItem);
+        return pos.equals(this.destroyBlockPos) && !destroyingItem.shouldCauseBlockBreakReset(selected);
     }
 
     private void ensureHasSentCarriedItem() {
@@ -443,6 +_,9 @@
         this.ensureHasSentCarriedItem();
         Vec3 location = hitResult.getLocation().subtract(entity.getX(), entity.getY(), entity.getZ());
         this.connection.send(new ServerboundInteractPacket(entity.getId(), hand, location, player.isShiftKeyDown()));
+        if (this.localPlayerMode == GameType.SPECTATOR) return InteractionResult.PASS; // don't fire for spectators to match non-specific EntityInteract
+        InteractionResult cancelResult = net.neoforged.neoforge.common.CommonHooks.onInteractEntityAt(player, entity, hitResult, hand);
+        if(cancelResult != null) return cancelResult;
         return this.localPlayerMode == GameType.SPECTATOR ? InteractionResult.PASS : player.interactOn(entity, hand, location);
     }
 
