--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -124,6 +124,7 @@
 import org.jspecify.annotations.Nullable;
 
 public abstract class Player extends Avatar implements ContainerUser, net.neoforged.neoforge.common.extensions.IPlayerExtension {
+    public static final String PERSISTED_NBT_TAG = "PlayerPersisted";
     public static final int MAX_HEALTH = 20;
     public static final int SLEEP_DURATION = 100;
     public static final int WAKE_UP_DURATION = 10;
@@ -684,7 +711,8 @@
                 return false;
             } else {
                 this.removeEntitiesOnShoulder();
-                if (source.scalesWithDifficulty()) {
+                damage = Math.max(0.0F, source.type().scaling().getScalingFunction().scaleDamage(source, this, damage, this.level().getDifficulty()));
+                if (false && source.scalesWithDifficulty()) {
                     if (level.getDifficulty() == Difficulty.PEACEFUL) {
                         damage = 0.0F;
                     }
@@ -742,11 +770,14 @@
     @Override
     protected void actuallyHurt(ServerLevel level, DamageSource source, float dmg) {
         if (!this.isInvulnerableTo(level, source)) {
-            dmg = this.getDamageAfterArmorAbsorb(source, dmg);
-            dmg = this.getDamageAfterMagicAbsorb(source, dmg);
-            float var8 = Math.max(dmg - this.getAbsorptionAmount(), 0.0F);
-            this.setAbsorptionAmount(this.getAbsorptionAmount() - (dmg - var8));
-            float absorbedDamage = dmg - var8;
+            this.damageContainers.peek().setReduction(net.neoforged.neoforge.common.damagesource.DamageContainer.Reduction.ARMOR, this.damageContainers.peek().getNewDamage() - this.getDamageAfterArmorAbsorb(source, this.damageContainers.peek().getNewDamage()));
+            this.getDamageAfterMagicAbsorb(source, this.damageContainers.peek().getNewDamage());
+            float damage = net.neoforged.neoforge.common.CommonHooks.onLivingDamagePre(this, this.damageContainers.peek());
+            this.damageContainers.peek().setReduction(net.neoforged.neoforge.common.damagesource.DamageContainer.Reduction.ABSORPTION, Math.min(this.getAbsorptionAmount(), damage));
+            float absorbed = Math.min(damage, this.damageContainers.peek().getReduction(net.neoforged.neoforge.common.damagesource.DamageContainer.Reduction.ABSORPTION));
+            this.setAbsorptionAmount(Math.max(0, this.getAbsorptionAmount() - absorbed));
+            float var8 = this.damageContainers.peek().getNewDamage();
+            float absorbedDamage = absorbed;
             if (absorbedDamage > 0.0F && absorbedDamage < 3.4028235E37F) {
                 this.awardStat(Stats.DAMAGE_ABSORBED, Math.round(absorbedDamage * 10.0F));
             }
@@ -971,7 +1023,8 @@
                     if (wasHurt) {
                         this.causeExtraKnockback(entity, this.getKnockback(entity, damageSource) + (knockbackAttack ? 0.5F : 0.0F), oldMovement);
                         if (sweepAttack) {
-                            this.doSweepAttack(entity, baseDamage, damageSource, attackStrengthScale);
+                            var sweepHitBox = attackingItemStack.getSweepHitBox(this, entity);
+                            this.doSweepAttack(entity, baseDamage, damageSource, attackStrengthScale, sweepHitBox);
                         }
 
                         this.attackVisualEffects(entity, criticalAttack, sweepAttack, fullStrengthAttack, false, magicBoost);
@@ -1066,11 +1122,12 @@
 
     private void itemAttackInteraction(Entity entity, ItemStack attackingItemStack, DamageSource damageSource, boolean applyToTarget) {
         Entity hurtTarget = entity;
-        if (entity instanceof EnderDragonPart) {
-            hurtTarget = ((EnderDragonPart)entity).parentMob;
+        if (entity instanceof net.neoforged.neoforge.entity.PartEntity<?> partEntity) {
+            hurtTarget = partEntity.getParent();
         }
 
         boolean itemHurtEnemy = false;
+        ItemStack copy = attackingItemStack.copy();
         if (this.level() instanceof ServerLevel serverLevel) {
             if (hurtTarget instanceof LivingEntity livingTarget) {
                 itemHurtEnemy = attackingItemStack.hurtEnemy(livingTarget, this);
@@ -1406,7 +1472,8 @@
 
     @Override
     public boolean causeFallDamage(double fallDistance, float damageModifier, DamageSource damageSource) {
-        if (this.abilities.mayfly) {
+        if (this.mayFly()) {
+            net.neoforged.neoforge.event.EventHooks.onPlayerFall(this, damageModifier, damageModifier);
             return false;
         } else {
             if (fallDistance >= 2.0) {
@@ -1843,21 +1922,21 @@
         if (!(heldWeapon.getItem() instanceof ProjectileWeaponItem)) {
             return ItemStack.EMPTY;
         } else {
-            Predicate<ItemStack> supportedProjectiles = ((ProjectileWeaponItem)heldWeapon.getItem()).getSupportedHeldProjectiles();
+            Predicate<ItemStack> supportedProjectiles = ((ProjectileWeaponItem)heldWeapon.getItem()).getSupportedHeldProjectiles(heldWeapon);
             ItemStack heldProjectile = ProjectileWeaponItem.getHeldProjectile(this, supportedProjectiles);
             if (!heldProjectile.isEmpty()) {
-                return heldProjectile;
+                return net.neoforged.neoforge.common.CommonHooks.getProjectile(this, heldWeapon, heldProjectile);
             } else {
-                supportedProjectiles = ((ProjectileWeaponItem)heldWeapon.getItem()).getAllSupportedProjectiles();
+                supportedProjectiles = ((ProjectileWeaponItem)heldWeapon.getItem()).getAllSupportedProjectiles(heldWeapon);
 
                 for (int i = 0; i < this.inventory.getContainerSize(); i++) {
                     ItemStack itemStack = this.inventory.getItem(i);
                     if (supportedProjectiles.test(itemStack)) {
-                        return itemStack;
+                        return net.neoforged.neoforge.common.CommonHooks.getProjectile(this, heldWeapon, itemStack);
                     }
                 }
 
-                return this.hasInfiniteMaterials() ? new ItemStack(Items.ARROW) : ItemStack.EMPTY;
+                return net.neoforged.neoforge.common.CommonHooks.getProjectile(this, heldWeapon, this.abilities.instabuild ? ((ProjectileWeaponItem)heldWeapon.getItem()).getDefaultCreativeAmmo(this, heldWeapon) : ItemStack.EMPTY);
             }
         }
     }
