--- a/net/minecraft/client/renderer/GameRenderer.java
+++ b/net/minecraft/client/renderer/GameRenderer.java
@@ -219,6 +_,7 @@
                 break;
             case null:
             default:
+                if (!net.neoforged.neoforge.client.ClientHooks.loadEntityShader(cameraEntity, this))
                 this.clearPostEffect();
         }
     }
@@ -308,6 +_,9 @@
             if (hurt < 0.0F) {
                 return;
             }
+
+            // Neo: Prevent screen shake if the damage type is marked as "forge:no_flinch"
+            if (cameraState.entityRenderState.preventFlinch) return;
 
             hurt /= cameraState.entityRenderState.hurtDuration;
             hurt = Mth.sin(hurt * hurt * hurt * hurt * (float) Math.PI);
