--- a/net/minecraft/client/renderer/LevelRenderer.java
+++ b/net/minecraft/client/renderer/LevelRenderer.java
@@ -225,11 +_,12 @@
                 deltaPartialTick,
                 this.levelRenderState.cloudColor,
                 this.levelRenderState.cloudHeight,
-                this.optionsRenderState.cloudRange
+                this.optionsRenderState.cloudRange,
+                modelViewMatrix
             );
         }
 
-        this.addWeatherPass(frame, terrainFog);
+        this.addWeatherPass(frame, terrainFog, modelViewMatrix);
         if (transparencyChain != null) {
             transparencyChain.addToFrame(frame, screenWidth, screenHeight, this.targets);
         }
@@ -452,6 +_,21 @@
         float cloudHeight,
         int cloudRange
     ) {
+        addCloudsPass(frame, cloudStatus, cameraPosition, gameTime, partialTicks, cloudColor, cloudHeight, cloudRange, RenderSystem.getModelViewMatrix());
+    }
+
+    // Neo: Add modelViewMatrix for use in our event
+    private void addCloudsPass(
+        FrameGraphBuilder frame,
+        CloudStatus cloudStatus,
+        Vec3 cameraPosition,
+        long gameTime,
+        float partialTicks,
+        int cloudColor,
+        float cloudHeight,
+        int cloudRange,
+        Matrix4fc modelViewMatrix
+    ) {
         FramePass pass = frame.addPass("clouds");
         if (this.targets.clouds != null) {
             this.targets.clouds = pass.readsAndWrites(this.targets.clouds);
@@ -459,10 +_,22 @@
             this.targets.main = pass.readsAndWrites(this.targets.main);
         }
 
-        pass.executes(() -> this.cloudRenderer.render(cloudColor, cloudStatus, cloudHeight, cloudRange, cameraPosition, gameTime, partialTicks));
+        pass.executes(() -> {
+            if (this.levelRenderState.customCloudsRenderer == null || !this.levelRenderState.customCloudsRenderer.renderClouds(this.levelRenderState, cameraPosition, cloudStatus, cloudColor, cloudHeight, cloudRange, modelViewMatrix)) {
+                this.cloudRenderer.render(cloudColor, cloudStatus, cloudHeight, cloudRange, cameraPosition, gameTime, partialTicks);
+            }
+        });
     }
 
+    /**
+     * @deprecated Neo: use {@link #addWeatherPass(FrameGraphBuilder, GpuBufferSlice, Matrix4fc)} instead
+     */
+    @Deprecated
     private void addWeatherPass(FrameGraphBuilder frame, GpuBufferSlice fog) {
+        addWeatherPass(frame, fog, RenderSystem.getModelViewMatrix());
+    }
+
+    private void addWeatherPass(FrameGraphBuilder frame, GpuBufferSlice fog, Matrix4fc modelViewMatrix) {
         int renderDistance = this.optionsRenderState.renderDistance * 16;
         FramePass pass = frame.addPass("weather");
         if (this.targets.weather != null) {
@@ -475,7 +_,10 @@
             () -> {
                 RenderSystem.setShaderFog(fog);
                 CameraRenderState cameraState = this.levelRenderState.cameraRenderState;
-                this.weatherEffectRenderer.render(cameraState.pos, this.levelRenderState.weatherRenderState);
+                this.weatherEffectRenderer.render(cameraState.pos, this.levelRenderState.weatherRenderState, this.levelRenderState);
+                Profiler.get().push("neoforge_render_after_weather");
+                net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.client.event.RenderLevelStageEvent.AfterWeather(this, this.levelRenderState, null, modelViewMatrix, this.getRenderableSections()));
+                Profiler.get().pop();
                 this.worldBorderRenderer
                     .render(this.levelRenderState.worldBorderRenderState, cameraState.pos, renderDistance, this.levelRenderState.cameraRenderState.depthFar);
             }
