--- a/net/minecraft/world/level/block/state/BlockBehaviour.java
+++ b/net/minecraft/world/level/block/state/BlockBehaviour.java
@@ -343,8 +342,8 @@
         if (destroySpeed == -1.0F) {
             return 0.0F;
         } else {
-            int modifier = player.hasCorrectToolForDrops(state) ? 30 : 100;
-            return player.getDestroySpeed(state) / destroySpeed / modifier;
+            int i = net.neoforged.neoforge.event.EventHooks.doPlayerHarvestCheck(player, state, level, pos) ? 30 : 100;
+            return player.getDestroySpeed(state, pos) / destroySpeed / (float)i;
         }
     }
 
