--- a/net/minecraft/world/level/gamerules/GameRuleType.java
+++ b/net/minecraft/world/level/gamerules/GameRuleType.java
@@ -2,6 +2,7 @@
 
 import net.minecraft.util.StringRepresentable;
 
+@net.neoforged.fml.common.asm.enumextension.NamedEnum
 public enum GameRuleType implements StringRepresentable, net.neoforged.fml.common.asm.enumextension.IExtensibleEnum {
     INT("integer"),
     BOOL("boolean");
