--- a/net/minecraft/world/level/block/piston/PistonBaseBlock.java
+++ b/net/minecraft/world/level/block/piston/PistonBaseBlock.java
@@ -162,6 +_,7 @@
 
         RandomSource random = level.getRandom();
         if (b0 == 0) {
+            if (net.neoforged.neoforge.event.EventHooks.onPistonMovePre(level, pos, direction, true)) return false;
             if (!this.moveBlocks(level, pos, direction, true)) {
                 return false;
             }
@@ -218,6 +_,7 @@
             level.gameEvent(GameEvent.BLOCK_DEACTIVATE, pos, GameEvent.Context.of(movingPistonState));
         }
 
+        net.neoforged.neoforge.event.EventHooks.onPistonMovePost(level, pos, direction, (b0 == 0));
         return true;
     }
 
@@ -367,6 +_,10 @@
     @Override
     protected BlockState rotate(BlockState state, Rotation rotation) {
         return state.setValue(FACING, rotation.rotate(state.getValue(FACING)));
+    }
+
+    public BlockState rotate(BlockState state, net.minecraft.world.level.LevelAccessor world, BlockPos pos, Rotation direction) {
+         return state.getValue(EXTENDED) ? state : super.rotate(state, world, pos, direction);
     }
 
     @Override
