--- a/net/minecraft/world/item/AxeItem.java
+++ b/net/minecraft/world/item/AxeItem.java
@@ -64,7 +66,7 @@
         if (playerHasBlockingItemUseIntent(context)) {
             return InteractionResult.PASS;
         } else {
-            Optional<BlockState> newBlock = this.evaluateNewBlockState(level, pos, player, level.getBlockState(pos));
+            Optional<BlockState> newBlock = this.evaluateNewBlockState(level, pos, player, level.getBlockState(pos), context);
             if (newBlock.isEmpty()) {
                 return InteractionResult.PASS;
             } else {
