--- a/net/minecraft/world/item/ItemDisplayContext.java
+++ b/net/minecraft/world/item/ItemDisplayContext.java
@@ -5,6 +5,9 @@
 import net.minecraft.util.ByIdMap;
 import net.minecraft.util.StringRepresentable;
 
+@net.neoforged.fml.common.asm.enumextension.IndexedEnum
+@net.neoforged.fml.common.asm.enumextension.NamedEnum(1)
+@net.neoforged.fml.common.asm.enumextension.NetworkedEnum(net.neoforged.fml.common.asm.enumextension.NetworkedEnum.NetworkCheck.CLIENTBOUND)
 public enum ItemDisplayContext implements StringRepresentable, net.neoforged.fml.common.asm.enumextension.IExtensibleEnum {
     NONE(0, "none"),
     THIRD_PERSON_LEFT_HAND(1, "thirdperson_lefthand"),
