--- a/net/minecraft/world/item/BucketItem.java
+++ b/net/minecraft/world/item/BucketItem.java
@@ -105,7 +_,12 @@
     }
 
     @Override
+    @Deprecated // Neo: use the ItemStack sensitive version
     public boolean emptyContents(@Nullable LivingEntity user, Level level, BlockPos pos, @Nullable BlockHitResult hitResult) {
+        return this.emptyContents(user, level, pos, hitResult, null);
+    }
+
+    public boolean emptyContents(@Nullable LivingEntity user, Level level, BlockPos pos, @Nullable BlockHitResult hitResult, @Nullable ItemStack containerItem) {
         if (!(this.content instanceof FlowingFluid flowingFluid)) {
             return false;
         } else {
@@ -132,7 +_,7 @@
                 }
 
                 return true;
-            } else if (block instanceof LiquidBlockContainer container && this.content == Fluids.WATER) {
+            } else if (block instanceof LiquidBlockContainer container && container.canPlaceLiquid(user, level, pos, blockState, content)) {
                 container.placeLiquid(level, pos, blockState, flowingFluid.getSource(false));
                 this.playEmptySound(user, level, pos);
                 return true;
@@ -152,9 +_,15 @@
     }
 
     protected void playEmptySound(@Nullable LivingEntity user, LevelAccessor level, BlockPos pos) {
-        SoundEvent soundEvent = this.content.is(FluidTags.LAVA) ? SoundEvents.BUCKET_EMPTY_LAVA : SoundEvents.BUCKET_EMPTY;
+        SoundEvent soundEvent = this.content.getFluidType().getSound(user, level, pos, net.neoforged.neoforge.common.SoundActions.BUCKET_EMPTY);
+        if(soundEvent == null) soundEvent = this.content.is(FluidTags.LAVA) ? SoundEvents.BUCKET_EMPTY_LAVA : SoundEvents.BUCKET_EMPTY;
         level.playSound(user, pos, soundEvent, SoundSource.BLOCKS, 1.0F, 1.0F);
         level.gameEvent(user, GameEvent.FLUID_PLACE, pos);
+    }
+
+    protected boolean canBlockContainFluid(@Nullable Player player, Level worldIn, BlockPos posIn, BlockState blockstate)
+    {
+        return blockstate.getBlock() instanceof LiquidBlockContainer && ((LiquidBlockContainer)blockstate.getBlock()).canPlaceLiquid(player, worldIn, posIn, blockstate, this.content);
     }
 
     public Fluid getContent() {
