--- a/net/minecraft/world/item/HoneycombItem.java
+++ b/net/minecraft/world/item/HoneycombItem.java
@@ -88,7 +_,7 @@
     }
 
     public static Optional<BlockState> getWaxed(BlockState oldState) {
-        return Optional.ofNullable(WAXABLES.get().get(oldState.getBlock())).map(b -> b.withPropertiesOf(oldState));
+        return Optional.ofNullable(net.neoforged.neoforge.common.DataMapHooks.getBlockWaxed(oldState.getBlock())).map(b -> b.withPropertiesOf(oldState));
     }
 
     @Override
