--- a/net/minecraft/world/item/ItemUseAnimation.java
+++ b/net/minecraft/world/item/ItemUseAnimation.java
@@ -50,6 +_,10 @@
         return this.name;
     }
 
+    public static net.neoforged.fml.common.asm.enumextension.ExtensionInfo getExtensionInfo() {
+        return net.neoforged.fml.common.asm.enumextension.ExtensionInfo.nonExtended(ItemUseAnimation.class);
+    }
+
     public boolean hasCustomArmTransform() {
         return this.customArmTransform;
     }
