--- a/net/minecraft/world/level/entity/TransientEntitySectionManager.java
+++ b/net/minecraft/world/level/entity/TransientEntitySectionManager.java
@@ -90,6 +91,7 @@
             Objects.requireNonNull(TransientEntitySectionManager.this);
             super();
             this.entity = entity;
+            this.realEntity = entity instanceof Entity ? (Entity) entity : null;
             this.currentSectionKey = currentSectionKey;
             this.currentSection = currentSection;
         }
