--- a/net/minecraft/network/CompressionEncoder.java
+++ b/net/minecraft/network/CompressionEncoder.java
@@ -17,6 +19,12 @@
 
     protected void encode(ChannelHandlerContext ctx, ByteBuf uncompressed, ByteBuf out) {
         int uncompressedLength = uncompressed.readableBytes();
+        if (!DISABLE_PACKET_DEBUG && uncompressedLength > net.minecraft.network.CompressionDecoder.MAXIMUM_UNCOMPRESSED_LENGTH) {
+            uncompressed.markReaderIndex();
+            LOGGER.error("Attempted to send packet over maximum protocol size: {} > {}\nData:\n{}", uncompressedLength, net.minecraft.network.CompressionDecoder.MAXIMUM_UNCOMPRESSED_LENGTH,
+                    net.neoforged.neoforge.logging.PacketDump.getContentDump(uncompressed));
+            uncompressed.resetReaderIndex();
+        }
         if (uncompressedLength > 8388608) {
             throw new IllegalArgumentException("Packet too big (is " + uncompressedLength + ", should be less than 8388608)");
         } else {
