--- a/net/minecraft/network/Connection.java
+++ b/net/minecraft/network/Connection.java
@@ -200,6 +211,7 @@
         if (protocol.flow() != this.getReceiving()) {
             throw new IllegalStateException("Invalid inbound protocol: " + protocol.id());
         } else {
+            this.inboundProtocol = protocol;
             this.packetListener = packetListener;
             this.disconnectListener = null;
             UnconfiguredPipelineHandler.InboundConfigurationTask configMessage = UnconfiguredPipelineHandler.setupInboundProtocol(protocol);
@@ -221,7 +233,12 @@
             BundlerInfo bundlerInfo = protocol.bundlerInfo();
             if (bundlerInfo != null) {
                 PacketBundleUnpacker newUnbundler = new PacketBundleUnpacker(bundlerInfo);
-                configMessage = configMessage.andThen(ctx -> ctx.pipeline().addAfter("encoder", "unbundler", newUnbundler));
+                configMessage = configMessage.andThen(ctx -> {
+                    ctx.pipeline().addAfter("encoder", "unbundler", newUnbundler);
+                    // Neo: our handlers must be between the encoder and the unbundler, so re-inject them
+                    // Note, this call must be inside the .andThen lambda, or it will actually run before the unbundler gets added.
+                    net.neoforged.neoforge.network.filters.NetworkFilters.injectIfNecessary(this);
+                });
             }
 
             boolean isLoginProtocol = protocol.id() == ConnectionProtocol.LOGIN;
@@ -590,5 +608,17 @@
 
     public void setBandwidthLogger(LocalSampleLogger bandwidthLogger) {
         this.bandwidthDebugMonitor = new BandwidthDebugMonitor(bandwidthLogger);
+    }
+
+    public Channel channel() {
+        return this.channel;
+    }
+
+    public PacketFlow getDirection() {
+        return this.receiving;
+    }
+
+    public ProtocolInfo<?> getInboundProtocol() {
+        return Objects.requireNonNull(this.inboundProtocol, "Inbound protocol not set?");
     }
 }
