--- a/net/minecraft/client/gui/screens/TitleScreen.java
+++ b/net/minecraft/client/gui/screens/TitleScreen.java
@@ -158,7 +164,7 @@
                 Button.builder(
                         Component.literal("Create Test World"), button -> CreateWorldScreen.testWorld(this.minecraft, () -> this.minecraft.setScreen(this))
                     )
-                    .bounds(this.width / 2 - 100, topPos += spacing, 200, 20)
+                    .bounds(this.width / 2 - 100, topPos += spacing, 98, 20)
                     .build()
             );
         }
