--- a/net/minecraft/client/gui/screens/worldselection/WorldSelectionList.java
+++ b/net/minecraft/client/gui/screens/worldselection/WorldSelectionList.java
@@ -515,6 +516,7 @@
             this.infoText.setPosition(textX, this.getContentY() + 9 + 9 + 3);
             this.infoText.extractRenderState(graphics, mouseX, mouseY, a);
             graphics.blit(RenderPipelines.GUI_TEXTURED, this.icon.textureLocation(), this.getContentX(), this.getContentY(), 0.0F, 0.0F, 32, 32, 32, 32);
+            renderExperimentalWarning(graphics, mouseX, mouseY, hovered);
             if (this.list.entryType == WorldSelectionList.EntryType.SINGLEPLAYER && (this.minecraft.options.touchscreen().get() || hovered)) {
                 graphics.fill(this.getContentX(), this.getContentY(), this.getContentX() + 32, this.getContentY() + 32, -1601138544);
                 int relX = mouseX - this.getContentX();
