--- a/net/minecraft/network/PacketProcessor.java
+++ b/net/minecraft/network/PacketProcessor.java
@@ -13,7 +_,7 @@
 
 public class PacketProcessor implements AutoCloseable {
     private static final Logger LOGGER = LogUtils.getLogger();
-    private final Queue<PacketProcessor.ListenerAndPacket<?>> packetsToBeHandled = Queues.newConcurrentLinkedQueue();
+    private final Queue<net.neoforged.neoforge.network.handling.QueuedPacket> packetsToBeHandled = Queues.newConcurrentLinkedQueue();
     private final Thread runningThread;
     private boolean closed;
 
