--- a/com/mojang/blaze3d/platform/TextureUtil.java
+++ b/com/mojang/blaze3d/platform/TextureUtil.java
@@ -80,6 +84,7 @@
                         };
                         case RGBA8 -> byteOffset -> data.getInt(byteOffset);
                         case DEPTH32 -> byteOffset -> ARGB.gray(data.getFloat(byteOffset));
+                        case DEPTH24_STENCIL8, DEPTH32_STENCIL8 -> _ -> -1;
                     };
                     int offsetx = 0;
 
