--- a/net/minecraft/client/renderer/texture/atlas/SpriteSourceList.java
+++ b/net/minecraft/client/renderer/texture/atlas/SpriteSourceList.java
@@ -58,7 +_,7 @@
                 }
             }
         };
-        this.sources.forEach(s -> s.run(resourceManager, output));
+        this.sources.forEach(s -> s.run(resourceManager, output, additionalMetadata));
         Builder<SpriteSource.Loader> result = ImmutableList.builder();
         result.add(loader -> MissingTextureAtlasSprite.create());
         result.addAll(sprites.values());
