--- a/net/minecraft/client/renderer/block/BlockModelRenderState.java
+++ b/net/minecraft/client/renderer/block/BlockModelRenderState.java
@@ -31,8 +_,10 @@
     private @Nullable IntList tintLayers;
     public int blockLightCoords;
     private @Nullable RandomSource randomSource;
+    private boolean hasTranslucency;
 
     public void clear() {
+        this.hasTranslucency = false;
         this.modelParts = null;
         this.transformation = null;
         this.renderType = null;
@@ -60,6 +_,7 @@
     public List<BlockStateModelPart> setupModel(Matrix4fc transformation, boolean hasTranslucency) {
         this.transformation = identityToNull(transformation);
         this.renderType = hasTranslucency ? Sheets.translucentBlockItemSheet() : Sheets.cutoutBlockItemSheet();
+        this.hasTranslucency = hasTranslucency;
         if (this.modelParts == null) {
             this.modelParts = new ObjectArrayList<>();
         } else {
