--- a/net/minecraft/client/renderer/block/FluidModel.java
+++ b/net/minecraft/client/renderer/block/FluidModel.java
@@ -28,7 +_,7 @@
                 transparency = transparency.or(getTransparency(overlayMaterial));
             }
 
-            return new FluidModel(ChunkSectionLayer.byTransparency(transparency), stillMaterial, flowingMaterial, overlayMaterial, this.tintSource);
+            return new FluidModel(ChunkSectionLayer.byTransparency(transparency), stillMaterial, flowingMaterial, overlayMaterial, this.fluidTintSource, this.customRenderer);
         }
 
         private static Transparency getTransparency(Material.Baked material) {
