--- a/net/minecraft/world/item/crafting/ShapedRecipePattern.java
+++ b/net/minecraft/world/item/crafting/ShapedRecipePattern.java
@@ -205,16 +228,16 @@
 
     public record Data(Map<Character, Ingredient> key, List<String> pattern) {
         private static final Codec<List<String>> PATTERN_CODEC = Codec.STRING.listOf().comapFlatMap(strings -> {
-            if (strings.size() > 3) {
-                return DataResult.error(() -> "Invalid pattern: too many rows, 3 is maximum");
+            if (strings.size() > maxHeight) {
+                return DataResult.error(() -> "Invalid pattern: too many rows, %s is maximum".formatted(maxHeight));
             } else if (strings.isEmpty()) {
                 return DataResult.error(() -> "Invalid pattern: empty pattern not allowed");
             } else {
                 int firstLength = strings.getFirst().length();
 
                 for (String line : strings) {
-                    if (line.length() > 3) {
-                        return DataResult.error(() -> "Invalid pattern: too many columns, 3 is maximum");
+                    if (line.length() > maxWidth) {
+                        return DataResult.error(() -> "Invalid pattern: too many columns, %s is maximum".formatted(maxWidth));
                     }
 
                     if (firstLength != line.length()) {
