--- a/net/minecraft/client/Options.java
+++ b/net/minecraft/client/Options.java
@@ -1427,6 +1428,7 @@
             value -> this.setGraphicsPresetToCustom()
         );
         this.syncWrites = Util.getPlatform() == Util.OS.WINDOWS;
+        net.neoforged.neoforge.client.ClientHooks.onRegisterKeyMappings(this, KeyMapping.Category.SORT_ORDER);
         this.load();
         this.initialExclusiveFullscreen = this.exclusiveFullscreen.get();
     }
