--- a/net/minecraft/CrashReportCategory.java
+++ b/net/minecraft/CrashReportCategory.java
@@ -181,16 +_,16 @@
 
         if (this.stackTrace.length > 0) {
             builder.append("\nStacktrace:");
-
-            for (StackTraceElement element : this.stackTrace) {
-                builder.append("\n\tat ");
-                builder.append(element);
-            }
+            builder.append(net.neoforged.neoforge.logging.CrashReportExtender.generateEnhancedStackTrace(this.stackTrace));
         }
     }
 
     public StackTraceElement[] getStacktrace() {
         return this.stackTrace;
+    }
+
+    public void setStackTrace(StackTraceElement[] stackTrace) {
+        this.stackTrace = stackTrace;
     }
 
     public static void populateBlockDetails(CrashReportCategory category, LevelHeightAccessor levelHeightAccessor, BlockPos pos, BlockState state) {
