--- a/net/minecraft/server/level/ServerPlayerGameMode.java
+++ b/net/minecraft/server/level/ServerPlayerGameMode.java
@@ -80,7 +_,10 @@
         this.previousGameModeForPlayer = previousGameModeForPlayer;
         this.gameModeForPlayer = gameModeForPlayer;
         Abilities abilities = this.player.getAbilities();
+        // Neo: preserve flying state, removed on tick if Attribute or ability no longer applies
+        boolean wasFlying = abilities.flying;
         gameModeForPlayer.updatePlayerAbilities(abilities);
+        abilities.flying = wasFlying || abilities.flying;
     }
 
     private boolean isInRangeOfGround() {
@@ -148,6 +_,10 @@
     }
 
     public void handleBlockBreakAction(BlockPos pos, ServerboundPlayerActionPacket.Action action, Direction direction, int maxY, int sequence) {
+        net.neoforged.neoforge.event.entity.player.PlayerInteractEvent.LeftClickBlock event = net.neoforged.neoforge.common.CommonHooks.onLeftClickBlock(player, pos, direction, action);
+        if (event.isCanceled()) {
+            return;
+        }
         if (!this.player.isWithinBlockInteractionRange(pos, 1.0)) {
             this.debugLogging(pos, false, sequence, "too far");
         } else if (pos.getY() > maxY) {
@@ -192,6 +_,7 @@
                         blockState,
                         item -> this.player.onEquippedItemBroken(item, EquipmentSlot.MAINHAND)
                     );
+                    if (event.getUseBlock() != net.minecraft.util.TriState.FALSE)
                     blockState.attack(this.level, pos, this.player);
                     progress = blockState.getDestroyProgress(this.player, this.player.level(), pos);
                 }
@@ -360,10 +_,15 @@
                 return InteractionResult.PASS;
             }
         } else {
+            UseOnContext context = new UseOnContext(player, hand, hitResult);
+            if (event.getUseItem() != net.minecraft.util.TriState.FALSE) {
+                InteractionResult result = itemStack.onItemUseFirst(context);
+                if (result != InteractionResult.PASS) return result;
+            }
             boolean haveSomethingInOurHands = !player.getMainHandItem().isEmpty() || !player.getOffhandItem().isEmpty();
-            boolean suppressUsingBlock = player.isSecondaryUseActive() && haveSomethingInOurHands;
+            boolean suppressUsingBlock = (player.isSecondaryUseActive() && haveSomethingInOurHands) && !(player.getMainHandItem().doesSneakBypassUse(level, pos, player) && player.getOffhandItem().doesSneakBypassUse(level, pos, player));
             ItemStack usedItemStack = itemStack.copy();
-            if (!suppressUsingBlock) {
+            if (event.getUseBlock().isTrue() || (event.getUseBlock().isDefault() && !suppressUsingBlock)) {
                 InteractionResult itemUse = state.useItemOn(player.getItemInHand(hand), level, player, hand, hitResult);
                 if (itemUse.consumesAction()) {
                     CriteriaTriggers.ITEM_USED_ON_BLOCK.trigger(player, pos, usedItemStack);
@@ -379,8 +_,8 @@
                 }
             }
 
-            if (!itemStack.isEmpty() && !player.getCooldowns().isOnCooldown(itemStack)) {
-                UseOnContext context = new UseOnContext(player, hand, hitResult);
+            if (event.getUseItem().isTrue() || (!itemStack.isEmpty() && !player.getCooldowns().isOnCooldown(itemStack))) {
+                if (event.getUseItem().isFalse()) return InteractionResult.PASS;
                 InteractionResult success;
                 if (player.hasInfiniteMaterials()) {
                     int count = itemStack.getCount();
