--- a/net/minecraft/server/level/ServerChunkCache.java
+++ b/net/minecraft/server/level/ServerChunkCache.java
@@ -101,7 +_,7 @@
             LOGGER.error("Failed to create dimension data storage directory", e);
         }
 
-        this.savedDataStorage = new SavedDataStorage(dataFolder, fixerUpper, level.registryAccess());
+        this.savedDataStorage = new SavedDataStorage(level, dataFolder, fixerUpper, level.registryAccess());
         this.ticketStorage = this.savedDataStorage.computeIfAbsent(TicketStorage.TYPE);
         this.chunkMap = new ChunkMap(
             level,
