--- a/net/minecraft/world/level/block/BaseRailBlock.java
+++ b/net/minecraft/world/level/block/BaseRailBlock.java
@@ -115,7 +115,7 @@
         if (level.isClientSide()) {
             return state;
         } else {
-            RailShape current = state.getValue(this.getShapeProperty());
+            RailShape current = getRailDirection(state, level, pos, null);
             return new RailState(level, pos, state).place(level.hasNeighborSignal(pos), first, current).getState();
         }
     }
