--- a/net/minecraft/world/level/block/Block.java
+++ b/net/minecraft/world/level/block/Block.java
@@ -291,12 +291,22 @@
         }
     }
 
+    /**
+     * @deprecated Neo: use overload with level context instead
+     */
+    @Deprecated
     public static boolean shouldRenderFace(BlockState state, BlockState neighborState, Direction direction) {
+        return shouldRenderFace(net.minecraft.world.level.EmptyBlockGetter.INSTANCE, BlockPos.ZERO, state, neighborState, direction);
+    }
+
+    public static boolean shouldRenderFace(BlockGetter level, BlockPos pos, BlockState state, BlockState neighborState, Direction direction) {
         VoxelShape occluder = neighborState.getFaceOcclusionShape(direction.getOpposite());
         if (occluder == Shapes.block()) {
             return false;
         } else if (state.skipRendering(neighborState, direction)) {
             return false;
+        } else if (neighborState.hidesNeighborFace(level, pos.relative(direction), state, direction.getOpposite()) && state.supportsExternalFaceHiding()) {
+            return false;
         } else if (occluder == Shapes.empty()) {
             return true;
         } else {
