--- a/net/minecraft/world/level/block/ComposterBlock.java
+++ b/net/minecraft/world/level/block/ComposterBlock.java
@@ -307,7 +318,7 @@
 
     private static BlockState addItem(@Nullable Entity sourceEntity, BlockState state, LevelAccessor level, BlockPos pos, ItemStack itemStack) {
         int fillLevel = state.getValue(LEVEL);
-        float chance = COMPOSTABLES.getFloat(itemStack.getItem());
+        float chance = getValue(itemStack);
         if ((fillLevel != 0 || !(chance > 0.0F)) && !(level.getRandom().nextDouble() < chance)) {
             return state;
         } else {
