--- a/net/minecraft/world/level/block/FlowerPotBlock.java
+++ b/net/minecraft/world/level/block/FlowerPotBlock.java
@@ -84,7 +106,7 @@
                 player.drop(plant, false);
             }
 
-            level.setBlock(pos, Blocks.FLOWER_POT.defaultBlockState(), 3);
+            level.setBlock(pos, getEmptyPot().defaultBlockState(), 3);
             level.gameEvent(player, GameEvent.BLOCK_CHANGE, pos);
             return InteractionResult.SUCCESS;
         }
