--- a/net/minecraft/world/level/block/PointedDripstoneBlock.java
+++ b/net/minecraft/world/level/block/PointedDripstoneBlock.java
@@ -552,7 +550,7 @@
     }
 
     private static boolean canFillCauldron(Fluid fluidAbove) {
-        return fluidAbove == Fluids.LAVA || fluidAbove == Fluids.WATER;
+        return fluidAbove.getFluidType().getDripInfo() != null;
     }
 
     private static boolean canGrow(BlockState rootState, BlockState aboveState) {
