--- a/net/minecraft/world/level/block/TntBlock.java
+++ b/net/minecraft/world/level/block/TntBlock.java
@@ -102,7 +106,7 @@
         if (!itemStack.is(Items.FLINT_AND_STEEL) && !itemStack.is(Items.FIRE_CHARGE)) {
             return super.useItemOn(itemStack, state, level, pos, player, hand, hitResult);
         } else {
-            if (prime(level, pos, player)) {
+            if (onCaughtFire(state, level, pos, hitResult.getDirection(), player)) {
                 level.setBlock(pos, Blocks.AIR.defaultBlockState(), 11);
                 Item item = itemStack.getItem();
                 if (itemStack.is(Items.FLINT_AND_STEEL)) {
