--- a/net/minecraft/client/resources/model/EquipmentClientInfo.java
+++ b/net/minecraft/client/resources/model/EquipmentClientInfo.java
@@ -110,6 +110,7 @@
     }
 
     @OnlyIn(Dist.CLIENT)
+    @net.neoforged.fml.common.asm.enumextension.NamedEnum
     public static enum LayerType implements StringRepresentable, net.neoforged.fml.common.asm.enumextension.IExtensibleEnum {
         HUMANOID("humanoid"),
         HUMANOID_LEGGINGS("humanoid_leggings"),
