--- a/net/minecraft/client/resources/model/ModelManager.java
+++ b/net/minecraft/client/resources/model/ModelManager.java
@@ -244,8 +287,9 @@
             }
 
             private Material.@Nullable Baked bake(Material material) {
-                Material.Baked itemMaterial = this.bakeForAtlas(material, itemAtlas);
-                return itemMaterial != null ? itemMaterial : this.bakeForAtlas(material, blockAtlas);
+                // Neo: fix missing block model rendering "random" texture (MC-307221)
+                Material.Baked blockMaterial = this.bakeForAtlas(material, blockAtlas);
+                return blockMaterial != null ? blockMaterial : this.bakeForAtlas(material, itemAtlas);
             }
 
             private Material.@Nullable Baked bakeForAtlas(Material material, SpriteLoader.Preparations atlas) {
