--- a/net/minecraft/client/resources/model/SimpleModelWrapper.java
+++ b/net/minecraft/client/resources/model/SimpleModelWrapper.java
@@ -43,7 +48,7 @@
         }
 
         if (forbiddenSprites != null) {
-            LOGGER.warn("Rejecting block model {}, since it contains sprites from outside of supported atlas: {}", location, forbiddenSprites);
+            LOGGER.warn("Rejecting block model {}, since it contains sprites from outside of supported atlas: {}", model.debugName(), forbiddenSprites);
             return modelBakery.missingBlockModelPart();
         } else {
             return new SimpleModelWrapper(geometry, hasAmbientOcclusion, particleMaterial);
