--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -1074,7 +1110,7 @@
             return BedBlock.findStandUpPosition(EntityType.PLAYER, level, pos, blockState.getValue(BedBlock.FACING), yaw)
                 .map(p -> ServerPlayer.RespawnPosAngle.of(p, pos, 0.0F));
         } else if (!forced) {
-            return Optional.empty();
+            return blockState.getRespawnPosition(EntityType.PLAYER, level, pos, yaw);
         } else {
             boolean freeBottom = block.isPossibleToRespawnInThis(blockState);
             BlockState topState = level.getBlockState(pos.above());
@@ -1096,6 +1132,7 @@
     }
 
     public @Nullable ServerPlayer teleport(TeleportTransition transition) {
+        if (!net.neoforged.neoforge.common.CommonHooks.onTravelToDimension(this, transition.newLevel().dimension())) return null;
         if (this.isRemoved()) {
             return null;
         } else {
@@ -1123,7 +1160,7 @@
                 PlayerList playerList = this.server.getPlayerList();
                 playerList.sendPlayerPermissionLevel(this);
                 oldLevel.removePlayerImmediately(this, Entity.RemovalReason.CHANGED_DIMENSION);
-                this.unsetRemoved();
+                this.revive();
                 ProfilerFiller profiler = Profiler.get();
                 profiler.push("moving");
                 if (lastDimension == Level.OVERWORLD && newLevel.dimension() == Level.NETHER) {
@@ -1143,11 +1180,15 @@
                 playerList.sendLevelInfo(this, newLevel);
                 playerList.sendAllPlayerInfo(this);
                 playerList.sendActivePlayerEffects(this);
+                // TODO 1.21: Play custom teleport sound
                 transition.postTeleportTransition().onTransition(this);
                 this.lastSentExp = -1;
                 this.lastSentHealth = -1.0F;
                 this.lastSentFood = -1;
                 this.teleportSpectators(transition, oldLevel);
+                // Neo: On the client all attachments are lost on dimension change and need to be re-sent
+                net.neoforged.neoforge.attachment.AttachmentSync.syncInitialPlayerAttachments(this);
+                net.neoforged.neoforge.event.EventHooks.firePlayerChangedDimensionEvent(this, lastDimension, transition.newLevel().dimension());
                 return this;
             }
         }
@@ -1232,6 +1284,23 @@
                         }
                     }
 
+        // End vanilla code
+                }
+            }
+        }
+
+        return Either.right(Unit.INSTANCE);
+        }).get();
+
+        // Fire the event. Return the error if one exists after the event, otherwise use the vanilla logic to start sleeping.
+        vanillaResult = net.neoforged.neoforge.event.EventHooks.canPlayerStartSleeping(this, pos, vanillaResult);
+        if (vanillaResult.left().isPresent()) {
+            return vanillaResult;
+        }
+
+        {
+            {{
+                // Start vanilla code
                     Either<Player.BedSleepingProblem, Unit> result = super.startSleepInBed(pos).ifRight(unit -> {
                         this.awardStat(Stats.SLEEP_IN_BED);
                         CriteriaTriggers.SLEPT_IN_BED.trigger(this);
@@ -1244,8 +1313,6 @@
                     return result;
                 }
             }
-        } else {
-            return Either.left(Player.BedSleepingProblem.OTHER_PROBLEM);
         }
     }
 
@@ -1339,11 +1407,19 @@
 
     @Override
     public OptionalInt openMenu(@Nullable MenuProvider provider) {
+        return openMenu(provider, (java.util.function.Consumer<net.minecraft.network.RegistryFriendlyByteBuf>) null);
+    }
+
+    @Override
+    public OptionalInt openMenu(@Nullable MenuProvider provider, java.util.function.@Nullable Consumer<net.minecraft.network.RegistryFriendlyByteBuf> extraDataWriter) {
         if (provider == null) {
             return OptionalInt.empty();
         } else {
             if (this.containerMenu != this.inventoryMenu) {
+                if (provider.shouldTriggerClientSideContainerClosingOnOpen())
                 this.closeContainer();
+                else
+                    this.doCloseContainer();
             }
 
             this.nextContainerCounter();
@@ -1355,9 +1431,31 @@
 
                 return OptionalInt.empty();
             } else {
+                // Neo: Support sending additional arbitrary data to menu factories on the client-side
+                var extraData = net.neoforged.neoforge.common.util.FriendlyByteBufUtil.writeCustomData(
+                        buffer -> {
+                            provider.writeClientSideData(menu, buffer);
+                            if (extraDataWriter != null) {
+                                extraDataWriter.accept(buffer);
+                            }
+                        },
+                        registryAccess()
+                );
+                if (extraData.length != 0) {
+                    this.connection.send(
+                            new net.neoforged.neoforge.network.payload.AdvancedOpenScreenPayload(
+                                    menu.containerId,
+                                    menu.getType(),
+                                    provider.getDisplayName(),
+                                    extraData
+                            )
+                    );
+                } else {
                 this.connection.send(new ClientboundOpenScreenPacket(menu.containerId, menu.getType(), provider.getDisplayName()));
+                }
                 this.initMenu(menu);
                 this.containerMenu = menu;
+                net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.entity.player.PlayerContainerEvent.Open(this, this.containerMenu));
                 return OptionalInt.of(this.containerCounter);
             }
         }
@@ -1739,6 +1848,8 @@
 
     public boolean setGameMode(GameType mode) {
         boolean wasSpectator = this.isSpectator();
+        mode = net.neoforged.neoforge.common.CommonHooks.onChangeGameType(this, this.gameMode.getGameModeForPlayer(), mode);
+        if (mode == null) return false;
         if (!this.gameMode.changeGameModeForPlayer(mode)) {
             return false;
         } else {
